// gill/dist/programs/token/index.node.cjs
'use strict';

var token2022 = require('@solana-program/token-2022');
var kit = require('@solana/kit');
var system = require('@solana-program/system');
var computeBudget = require('@solana-program/compute-budget');
var codecs = require('@solana/codecs');

// src/programs/token/addresses.ts
function checkedAddress(input) {
  return typeof input == "string" ? input : input.address;
}
function checkedTransactionSigner(input) {
  if (typeof input === "string" || "address" in input == false) input = kit.createNoopSigner(input);
  if (!kit.isTransactionSigner(input)) throw new Error("A signer or address is required");
  kit.assertIsTransactionSigner(input);
  return input;
}

// src/programs/token/addresses.ts
var TOKEN_PROGRAM_ADDRESS = "TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA";
async function getAssociatedTokenAccountAddress(mint, owner, tokenProgram) {
  return (await token2022.findAssociatedTokenPda({
    mint: checkedAddress(mint),
    owner: checkedAddress(owner),
    tokenProgram: checkedTokenProgramAddress(tokenProgram)
  }))[0];
}
function parseTokenProgramAddressOrMoniker(tokenProgram) {
  if (!kit.isAddress(tokenProgram)) {
    tokenProgram = tokenProgram.toLowerCase();
  }
  switch (tokenProgram) {
    case "legacy":
    case "token":
    case TOKEN_PROGRAM_ADDRESS: {
      return TOKEN_PROGRAM_ADDRESS;
    }
    case "token22":
    case "token2022":
    case "tokenextension":
    case "tokenextensions":
    case token2022.TOKEN_2022_PROGRAM_ADDRESS: {
      return token2022.TOKEN_2022_PROGRAM_ADDRESS;
    }
    default:
      throw Error("Unsupported token program. Try 'TOKEN_PROGRAM_ADDRESS' or 'TOKEN_2022_PROGRAM_ADDRESS'");
  }
}
function assertIsSupportedTokenProgram(tokenProgram) {
  if (tokenProgram !== TOKEN_PROGRAM_ADDRESS && tokenProgram !== token2022.TOKEN_2022_PROGRAM_ADDRESS) {
    throw Error("Unsupported token program. Try 'TOKEN_PROGRAM_ADDRESS' or 'TOKEN_2022_PROGRAM_ADDRESS'");
  }
}
function checkedTokenProgramAddress(tokenProgram) {
  if (!tokenProgram) return TOKEN_PROGRAM_ADDRESS;
  tokenProgram = parseTokenProgramAddressOrMoniker(tokenProgram);
  assertIsSupportedTokenProgram(tokenProgram);
  return tokenProgram;
}
function createTransaction({
  version,
  feePayer,
  instructions,
  latestBlockhash,
  computeUnitLimit,
  computeUnitPrice
}) {
  return kit.pipe(
    kit.createTransactionMessage({ version }),
    (tx) => {
      const withLifetime = latestBlockhash ? kit.setTransactionMessageLifetimeUsingBlockhash(latestBlockhash, tx) : tx;
      if (typeof feePayer !== "string" && "address" in feePayer && kit.isTransactionSigner(feePayer)) {
        return kit.setTransactionMessageFeePayerSigner(feePayer, withLifetime);
      } else return kit.setTransactionMessageFeePayer(feePayer, withLifetime);
    },
    (tx) => {
      const withComputeLimit = typeof computeUnitLimit !== "undefined" ? kit.appendTransactionMessageInstruction(
        computeBudget.getSetComputeUnitLimitInstruction({ units: Number(computeUnitLimit) }),
        tx
      ) : tx;
      const withComputePrice = typeof computeUnitPrice !== "undefined" ? kit.appendTransactionMessageInstruction(
        computeBudget.getSetComputeUnitPriceInstruction({ microLamports: Number(computeUnitPrice) }),
        withComputeLimit
      ) : withComputeLimit;
      return kit.appendTransactionMessageInstructions(instructions, withComputePrice);
    }
  );
}

// src/core/accounts.ts
function getMinimumBalanceForRentExemption(space = 0) {
  const RENT = {
    /**
     * Account storage overhead for calculation of base rent. (aka the number of bytes required to store an account with no data.
     */
    ACCOUNT_STORAGE_OVERHEAD: 128n,
    /**
     * Amount of time (in years) a balance must include rent for the account to
     * be rent exempt.
     */
    DEFAULT_EXEMPTION_THRESHOLD: BigInt(Math.floor(2 * 1e3)) / 1000n,
    /**
     * Default rental rate in lamports/byte-year. This calculation is based on:
     * - 10^9 lamports per SOL
     * - $1 per SOL
     * - $0.01 per megabyte day
     * - $3.65 per megabyte year
     */
    DEFAULT_LAMPORTS_PER_BYTE_YEAR: BigInt(
      Math.floor(1e9 / 100 * 365 / (1024 * 1024))
    )
  };
  return (RENT.ACCOUNT_STORAGE_OVERHEAD + BigInt(space)) * RENT.DEFAULT_LAMPORTS_PER_BYTE_YEAR * RENT.DEFAULT_EXEMPTION_THRESHOLD / 1n;
}
function getCollectionEncoder() {
  return kit.getStructEncoder([
    ["verified", kit.getBooleanEncoder()],
    ["key", kit.getAddressEncoder()]
  ]);
}
function getCollectionDetailsEncoder() {
  return kit.getDiscriminatedUnionEncoder([
    ["V1", kit.getStructEncoder([["size", kit.getU64Encoder()]])],
    [
      "V2",
      kit.getStructEncoder([["padding", kit.fixEncoderSize(kit.getBytesEncoder(), 8)]])
    ]
  ]);
}
function getCreatorEncoder() {
  return kit.getStructEncoder([
    ["address", kit.getAddressEncoder()],
    ["verified", kit.getBooleanEncoder()],
    ["share", kit.getU8Encoder()]
  ]);
}
function getDataV2Encoder() {
  return kit.getStructEncoder([
    ["name", kit.addEncoderSizePrefix(kit.getUtf8Encoder(), kit.getU32Encoder())],
    ["symbol", kit.addEncoderSizePrefix(kit.getUtf8Encoder(), kit.getU32Encoder())],
    ["uri", kit.addEncoderSizePrefix(kit.getUtf8Encoder(), kit.getU32Encoder())],
    ["sellerFeeBasisPoints", kit.getU16Encoder()],
    ["creators", kit.getOptionEncoder(kit.getArrayEncoder(getCreatorEncoder()))],
    ["collection", kit.getOptionEncoder(getCollectionEncoder())],
    ["uses", kit.getOptionEncoder(getUsesEncoder())]
  ]);
}
function getUsesEncoder() {
  return kit.getStructEncoder([
    ["useMethod", getUseMethodEncoder()],
    ["remaining", kit.getU64Encoder()],
    ["total", kit.getU64Encoder()]
  ]);
}
var UseMethod = /* @__PURE__ */ ((UseMethod2) => {
  UseMethod2[UseMethod2["Burn"] = 0] = "Burn";
  UseMethod2[UseMethod2["Multiple"] = 1] = "Multiple";
  UseMethod2[UseMethod2["Single"] = 2] = "Single";
  return UseMethod2;
})(UseMethod || {});
function getUseMethodEncoder() {
  return kit.getEnumEncoder(UseMethod);
}

// src/programs/token-metadata/generated/programs/tokenMetadata.ts
var TOKEN_METADATA_PROGRAM_ADDRESS = "metaqbxxUerdq28cj1RbAWkYQm3ybzjb6a8bt518x1s";
function expectAddress(value) {
  if (!value) {
    throw new Error("Expected a Address.");
  }
  if (typeof value === "object" && "address" in value) {
    return value.address;
  }
  if (Array.isArray(value)) {
    return value[0];
  }
  return value;
}
function getAccountMetaFactory(programAddress, optionalAccountStrategy) {
  return (account) => {
    if (!account.value) {
      return;
    }
    const writableRole = account.isWritable ? kit.AccountRole.WRITABLE : kit.AccountRole.READONLY;
    return Object.freeze({
      address: expectAddress(account.value),
      role: isTransactionSigner3(account.value) ? kit.upgradeRoleToSigner(writableRole) : writableRole,
      ...isTransactionSigner3(account.value) ? { signer: account.value } : {}
    });
  };
}
function isTransactionSigner3(value) {
  return !!value && typeof value === "object" && "address" in value && kit.isTransactionSigner(value);
}

// src/programs/token-metadata/generated/instructions/createMetadataAccountV3.ts
var CREATE_METADATA_ACCOUNT_V3_DISCRIMINATOR = 33;
function getCreateMetadataAccountV3InstructionDataEncoder() {
  return codecs.transformEncoder(
    codecs.getStructEncoder([
      ["discriminator", codecs.getU8Encoder()],
      ["data", getDataV2Encoder()],
      ["isMutable", codecs.getBooleanEncoder()],
      ["collectionDetails", codecs.getOptionEncoder(getCollectionDetailsEncoder())]
    ]),
    (value) => ({
      ...value,
      discriminator: CREATE_METADATA_ACCOUNT_V3_DISCRIMINATOR
    })
  );
}
function getCreateMetadataAccountV3Instruction(input, config) {
  const programAddress = config?.programAddress ?? TOKEN_METADATA_PROGRAM_ADDRESS;
  const originalAccounts = {
    metadata: { value: input.metadata ?? null, isWritable: true },
    mint: { value: input.mint ?? null, isWritable: false },
    mintAuthority: { value: input.mintAuthority ?? null, isWritable: false },
    payer: { value: input.payer ?? null, isWritable: true },
    updateAuthority: {
      value: input.updateAuthority ?? null,
      isWritable: false
    },
    systemProgram: { value: input.systemProgram ?? null, isWritable: false },
    rent: { value: input.rent ?? null, isWritable: false }
  };
  const accounts = originalAccounts;
  const args = { ...input };
  if (!accounts.systemProgram.value) {
    accounts.systemProgram.value = "11111111111111111111111111111111";
  }
  const getAccountMeta = getAccountMetaFactory();
  const instruction = {
    accounts: [
      getAccountMeta(accounts.metadata),
      getAccountMeta(accounts.mint),
      getAccountMeta(accounts.mintAuthority),
      getAccountMeta(accounts.payer),
      getAccountMeta(accounts.updateAuthority),
      getAccountMeta(accounts.systemProgram),
      getAccountMeta(accounts.rent)
    ].filter((x) => x !== void 0),
    programAddress,
    data: getCreateMetadataAccountV3InstructionDataEncoder().encode(args)
  };
  return instruction;
}
async function getTokenMetadataAddress(mint) {
  return (await kit.getProgramDerivedAddress({
    programAddress: TOKEN_METADATA_PROGRAM_ADDRESS,
    seeds: [
      Buffer.from("metadata"),
      kit.getAddressEncoder().encode(TOKEN_METADATA_PROGRAM_ADDRESS),
      kit.getAddressEncoder().encode(checkedAddress(mint))
    ]
  }))[0];
}
function getCreateTokenInstructions(args) {
  args.tokenProgram = checkedTokenProgramAddress(args.tokenProgram);
  args.feePayer = checkedTransactionSigner(args.feePayer);
  if (args.decimals == null) args.decimals = 9;
  if (!args.mintAuthority) args.mintAuthority = args.feePayer;
  if (!args.updateAuthority) args.updateAuthority = args.feePayer;
  if (args.freezeAuthority) args.freezeAuthority = checkedAddress(args.freezeAuthority);
  if (args.tokenProgram === token2022.TOKEN_2022_PROGRAM_ADDRESS) {
    const metadataPointer = token2022.extension("MetadataPointer", {
      metadataAddress: args.mint.address,
      authority: args.updateAuthority.address
    });
    const metadataExtensionData = token2022.extension("TokenMetadata", {
      updateAuthority: args.updateAuthority.address,
      mint: args.mint.address,
      name: args.metadata.name,
      symbol: args.metadata.symbol,
      uri: args.metadata.uri,
      // todo: support token22 additional metadata
      additionalMetadata: /* @__PURE__ */ new Map()
    });
    return [
      system.getCreateAccountInstruction({
        payer: args.feePayer,
        newAccount: args.mint,
        /**
         * token22 requires only the pre-mint-initialization extensions (like metadata pointer)
         * to be the `space`. then it will extend the account's space for each applicable extension
         * */
        space: BigInt(token2022.getMintSize([metadataPointer])),
        /**
         * token22 requires the total lamport balance for all extensions,
         * including pre-initialization and post-initialization
         */
        lamports: getMinimumBalanceForRentExemption(BigInt(token2022.getMintSize([metadataPointer, metadataExtensionData]))),
        programAddress: args.tokenProgram
      }),
      token2022.getInitializeMetadataPointerInstruction({
        authority: args.mintAuthority.address,
        metadataAddress: args.metadataAddress,
        mint: args.mint.address
      }),
      token2022.getInitializeMintInstruction(
        {
          mint: args.mint.address,
          decimals: Number(args.decimals),
          mintAuthority: args.mintAuthority.address,
          freezeAuthority: args.freezeAuthority || null
        },
        {
          programAddress: args.tokenProgram
        }
      ),
      token2022.getInitializeTokenMetadataInstruction({
        metadata: args.mint.address,
        mint: args.mint.address,
        mintAuthority: args.mintAuthority,
        name: args.metadata.name,
        symbol: args.metadata.symbol,
        uri: args.metadata.uri,
        updateAuthority: args.updateAuthority.address
      })
      // todo: support token22 additional metadata by adding that instruction(s) here
    ];
  } else {
    const space = BigInt(token2022.getMintSize());
    return [
      system.getCreateAccountInstruction({
        payer: args.feePayer,
        newAccount: args.mint,
        lamports: getMinimumBalanceForRentExemption(space),
        space,
        programAddress: args.tokenProgram
      }),
      token2022.getInitializeMintInstruction(
        {
          mint: args.mint.address,
          decimals: Number(args.decimals),
          mintAuthority: args.mintAuthority.address,
          freezeAuthority: args.freezeAuthority || null
        },
        {
          programAddress: args.tokenProgram
        }
      ),
      getCreateMetadataAccountV3Instruction({
        metadata: args.metadataAddress,
        mint: args.mint.address,
        mintAuthority: args.mintAuthority,
        payer: args.feePayer,
        updateAuthority: args.updateAuthority,
        data: {
          name: args.metadata.name,
          symbol: args.metadata.symbol,
          uri: args.metadata.uri,
          sellerFeeBasisPoints: 0,
          creators: null,
          collection: null,
          uses: null
        },
        isMutable: args.metadata.isMutable,
        collectionDetails: null
      })
    ];
  }
}
function getMintTokensInstructions(args) {
  args.tokenProgram = checkedTokenProgramAddress(args.tokenProgram);
  args.feePayer = checkedTransactionSigner(args.feePayer);
  args.mint = checkedAddress(args.mint);
  return [
    // create idempotent will gracefully fail if the ata already exists. this is the gold standard!
    token2022.getCreateAssociatedTokenIdempotentInstruction({
      owner: checkedAddress(args.destination),
      mint: args.mint,
      ata: args.ata,
      payer: args.feePayer,
      tokenProgram: args.tokenProgram
    }),
    token2022.getMintToInstruction(
      {
        mint: args.mint,
        mintAuthority: args.mintAuthority,
        token: args.ata,
        amount: args.amount
      },
      {
        programAddress: args.tokenProgram
      }
    )
  ];
}
function getTransferTokensInstructions(args) {
  args.tokenProgram = checkedTokenProgramAddress(args.tokenProgram);
  args.feePayer = checkedTransactionSigner(args.feePayer);
  args.mint = checkedAddress(args.mint);
  return [
    // create idempotent will gracefully fail if the ata already exists. this is the gold standard!
    token2022.getCreateAssociatedTokenIdempotentInstruction({
      owner: checkedAddress(args.destination),
      mint: args.mint,
      ata: args.destinationAta,
      payer: args.feePayer,
      tokenProgram: args.tokenProgram
    }),
    token2022.getTransferInstruction(
      {
        authority: args.authority,
        source: args.sourceAta,
        destination: args.destinationAta,
        amount: args.amount
      },
      {
        programAddress: args.tokenProgram
      }
    )
  ];
}
async function buildCreateTokenTransaction(args) {
  args.tokenProgram = checkedTokenProgramAddress(args.tokenProgram);
  args.feePayer = checkedTransactionSigner(args.feePayer);
  let metadataAddress = args.mint.address;
  if (args.tokenProgram === TOKEN_PROGRAM_ADDRESS) {
    metadataAddress = await getTokenMetadataAddress(args.mint);
    if (!args.computeUnitLimit) {
      args.computeUnitLimit = 6e4;
    }
  } else if (args.tokenProgram === token2022.TOKEN_2022_PROGRAM_ADDRESS) {
    if (!args.computeUnitLimit) {
      args.computeUnitLimit = 1e4;
    }
  }
  return createTransaction(
    (({ feePayer, version, computeUnitLimit, computeUnitPrice, latestBlockhash }) => ({
      feePayer,
      version: version || "legacy",
      computeUnitLimit,
      computeUnitPrice,
      latestBlockhash,
      instructions: getCreateTokenInstructions(
        (({
          decimals,
          mintAuthority,
          freezeAuthority,
          updateAuthority,
          metadata,
          feePayer: feePayer2,
          tokenProgram,
          mint
        }) => ({
          mint,
          feePayer: feePayer2,
          metadataAddress,
          metadata,
          decimals,
          mintAuthority,
          freezeAuthority,
          updateAuthority,
          tokenProgram
        }))(args)
      )
    }))(args)
  );
}

// src/programs/token/transactions/mint-tokens.ts
async function buildMintTokensTransaction(args) {
  args.tokenProgram = checkedTokenProgramAddress(args.tokenProgram);
  args.feePayer = checkedTransactionSigner(args.feePayer);
  args.mint = checkedAddress(args.mint);
  if (!args.ata) {
    args.ata = await getAssociatedTokenAccountAddress(args.mint, args.destination, args.tokenProgram);
  }
  if (!args.computeUnitLimit) {
    args.computeUnitLimit = 31e3;
  }
  return createTransaction(
    (({ feePayer, version, computeUnitLimit, computeUnitPrice, latestBlockhash }) => ({
      feePayer,
      version: version || "legacy",
      computeUnitLimit,
      computeUnitPrice,
      latestBlockhash,
      instructions: getMintTokensInstructions(
        (({ tokenProgram, feePayer: feePayer2, mint, ata, mintAuthority, amount, destination }) => ({
          tokenProgram,
          feePayer: feePayer2,
          mint,
          mintAuthority,
          ata,
          amount,
          destination
        }))(args)
      )
    }))(args)
  );
}

// src/programs/token/transactions/transfer-tokens.ts
async function buildTransferTokensTransaction(args) {
  args.tokenProgram = checkedTokenProgramAddress(args.tokenProgram);
  args.feePayer = checkedTransactionSigner(args.feePayer);
  args.mint = checkedAddress(args.mint);
  [args.destinationAta, args.sourceAta] = await Promise.all([
    !args.destinationAta ? getAssociatedTokenAccountAddress(args.mint, args.destination, args.tokenProgram) : args.destinationAta,
    !args.sourceAta ? getAssociatedTokenAccountAddress(args.mint, args.authority, args.tokenProgram) : args.sourceAta
  ]);
  if (!args.computeUnitLimit) {
    args.computeUnitLimit = 31e3;
  }
  return createTransaction(
    (({ feePayer, version, computeUnitLimit, computeUnitPrice, latestBlockhash }) => ({
      feePayer,
      version: version || "legacy",
      computeUnitLimit,
      computeUnitPrice,
      latestBlockhash,
      instructions: getTransferTokensInstructions(
        (({
          tokenProgram,
          feePayer: feePayer2,
          mint,
          amount,
          destination,
          authority,
          destinationAta,
          sourceAta
        }) => ({
          tokenProgram,
          feePayer: feePayer2,
          mint,
          amount,
          authority,
          destination,
          sourceAta,
          destinationAta
        }))(args)
      )
    }))(args)
  );
}

// src/programs/token/ui-amount.ts
function tokenUiAmountToAmount(uiAmount, decimals) {
  const uiAmountStr = uiAmount.toFixed(decimals);
  const decimalPointIndex = uiAmountStr.indexOf(".");
  let wholeNumber;
  let fractionalPart;
  if (decimalPointIndex === -1) {
    wholeNumber = uiAmountStr;
    fractionalPart = "";
  } else {
    wholeNumber = uiAmountStr.substring(0, decimalPointIndex);
    fractionalPart = uiAmountStr.substring(decimalPointIndex + 1);
  }
  if (fractionalPart.length > decimals) {
    fractionalPart = fractionalPart.substring(0, decimals);
  } else {
    fractionalPart = fractionalPart.padEnd(decimals, "0");
  }
  const rawAmountStr = wholeNumber + fractionalPart;
  if (rawAmountStr.replace(/0/g, "") === "") {
    return BigInt(0);
  }
  return BigInt(rawAmountStr);
}
function tokenAmountToUiAmount(amount, decimals) {
  let amountStr = amount.toString();
  if (amountStr.length <= decimals) {
    amountStr = amountStr.padStart(decimals + 1, "0");
  }
  const wholePartLength = amountStr.length - decimals;
  const wholePart = amountStr.substring(0, wholePartLength);
  const fractionalPart = amountStr.substring(wholePartLength);
  const uiAmountStr = `${wholePart || "0"}.${fractionalPart}`;
  return parseFloat(uiAmountStr);
}

Object.defineProperty(exports, "AMOUNT_TO_UI_AMOUNT_DISCRIMINATOR", {
  enumerable: true,
  get: function () { return token2022.AMOUNT_TO_UI_AMOUNT_DISCRIMINATOR; }
});
Object.defineProperty(exports, "APPLY_CONFIDENTIAL_PENDING_BALANCE_CONFIDENTIAL_TRANSFER_DISCRIMINATOR", {
  enumerable: true,
  get: function () { return token2022.APPLY_CONFIDENTIAL_PENDING_BALANCE_CONFIDENTIAL_TRANSFER_DISCRIMINATOR; }
});
Object.defineProperty(exports, "APPLY_CONFIDENTIAL_PENDING_BALANCE_DISCRIMINATOR", {
  enumerable: true,
  get: function () { return token2022.APPLY_CONFIDENTIAL_PENDING_BALANCE_DISCRIMINATOR; }
});
Object.defineProperty(exports, "APPROVE_CHECKED_DISCRIMINATOR", {
  enumerable: true,
  get: function () { return token2022.APPROVE_CHECKED_DISCRIMINATOR; }
});
Object.defineProperty(exports, "APPROVE_CONFIDENTIAL_TRANSFER_ACCOUNT_CONFIDENTIAL_TRANSFER_DISCRIMINATOR", {
  enumerable: true,
  get: function () { return token2022.APPROVE_CONFIDENTIAL_TRANSFER_ACCOUNT_CONFIDENTIAL_TRANSFER_DISCRIMINATOR; }
});
Object.defineProperty(exports, "APPROVE_CONFIDENTIAL_TRANSFER_ACCOUNT_DISCRIMINATOR", {
  enumerable: true,
  get: function () { return token2022.APPROVE_CONFIDENTIAL_TRANSFER_ACCOUNT_DISCRIMINATOR; }
});
Object.defineProperty(exports, "APPROVE_DISCRIMINATOR", {
  enumerable: true,
  get: function () { return token2022.APPROVE_DISCRIMINATOR; }
});
Object.defineProperty(exports, "ASSOCIATED_TOKEN_ERROR__INVALID_OWNER", {
  enumerable: true,
  get: function () { return token2022.ASSOCIATED_TOKEN_ERROR__INVALID_OWNER; }
});
Object.defineProperty(exports, "ASSOCIATED_TOKEN_PROGRAM_ADDRESS", {
  enumerable: true,
  get: function () { return token2022.ASSOCIATED_TOKEN_PROGRAM_ADDRESS; }
});
Object.defineProperty(exports, "AccountState", {
  enumerable: true,
  get: function () { return token2022.AccountState; }
});
Object.defineProperty(exports, "AssociatedTokenInstruction", {
  enumerable: true,
  get: function () { return token2022.AssociatedTokenInstruction; }
});
Object.defineProperty(exports, "AuthorityType", {
  enumerable: true,
  get: function () { return token2022.AuthorityType; }
});
Object.defineProperty(exports, "BURN_CHECKED_DISCRIMINATOR", {
  enumerable: true,
  get: function () { return token2022.BURN_CHECKED_DISCRIMINATOR; }
});
Object.defineProperty(exports, "BURN_DISCRIMINATOR", {
  enumerable: true,
  get: function () { return token2022.BURN_DISCRIMINATOR; }
});
Object.defineProperty(exports, "CLOSE_ACCOUNT_DISCRIMINATOR", {
  enumerable: true,
  get: function () { return token2022.CLOSE_ACCOUNT_DISCRIMINATOR; }
});
Object.defineProperty(exports, "CONFIDENTIAL_DEPOSIT_CONFIDENTIAL_TRANSFER_DISCRIMINATOR", {
  enumerable: true,
  get: function () { return token2022.CONFIDENTIAL_DEPOSIT_CONFIDENTIAL_TRANSFER_DISCRIMINATOR; }
});
Object.defineProperty(exports, "CONFIDENTIAL_DEPOSIT_DISCRIMINATOR", {
  enumerable: true,
  get: function () { return token2022.CONFIDENTIAL_DEPOSIT_DISCRIMINATOR; }
});
Object.defineProperty(exports, "CONFIDENTIAL_TRANSFER_CONFIDENTIAL_TRANSFER_DISCRIMINATOR", {
  enumerable: true,
  get: function () { return token2022.CONFIDENTIAL_TRANSFER_CONFIDENTIAL_TRANSFER_DISCRIMINATOR; }
});
Object.defineProperty(exports, "CONFIDENTIAL_TRANSFER_DISCRIMINATOR", {
  enumerable: true,
  get: function () { return token2022.CONFIDENTIAL_TRANSFER_DISCRIMINATOR; }
});
Object.defineProperty(exports, "CONFIDENTIAL_TRANSFER_WITH_FEE_CONFIDENTIAL_TRANSFER_DISCRIMINATOR", {
  enumerable: true,
  get: function () { return token2022.CONFIDENTIAL_TRANSFER_WITH_FEE_CONFIDENTIAL_TRANSFER_DISCRIMINATOR; }
});
Object.defineProperty(exports, "CONFIDENTIAL_TRANSFER_WITH_FEE_DISCRIMINATOR", {
  enumerable: true,
  get: function () { return token2022.CONFIDENTIAL_TRANSFER_WITH_FEE_DISCRIMINATOR; }
});
Object.defineProperty(exports, "CONFIDENTIAL_WITHDRAW_CONFIDENTIAL_TRANSFER_DISCRIMINATOR", {
  enumerable: true,
  get: function () { return token2022.CONFIDENTIAL_WITHDRAW_CONFIDENTIAL_TRANSFER_DISCRIMINATOR; }
});
Object.defineProperty(exports, "CONFIDENTIAL_WITHDRAW_DISCRIMINATOR", {
  enumerable: true,
  get: function () { return token2022.CONFIDENTIAL_WITHDRAW_DISCRIMINATOR; }
});
Object.defineProperty(exports, "CONFIGURE_CONFIDENTIAL_TRANSFER_ACCOUNT_CONFIDENTIAL_TRANSFER_DISCRIMINATOR", {
  enumerable: true,
  get: function () { return token2022.CONFIGURE_CONFIDENTIAL_TRANSFER_ACCOUNT_CONFIDENTIAL_TRANSFER_DISCRIMINATOR; }
});
Object.defineProperty(exports, "CONFIGURE_CONFIDENTIAL_TRANSFER_ACCOUNT_DISCRIMINATOR", {
  enumerable: true,
  get: function () { return token2022.CONFIGURE_CONFIDENTIAL_TRANSFER_ACCOUNT_DISCRIMINATOR; }
});
Object.defineProperty(exports, "CREATE_ASSOCIATED_TOKEN_DISCRIMINATOR", {
  enumerable: true,
  get: function () { return token2022.CREATE_ASSOCIATED_TOKEN_DISCRIMINATOR; }
});
Object.defineProperty(exports, "CREATE_ASSOCIATED_TOKEN_IDEMPOTENT_DISCRIMINATOR", {
  enumerable: true,
  get: function () { return token2022.CREATE_ASSOCIATED_TOKEN_IDEMPOTENT_DISCRIMINATOR; }
});
Object.defineProperty(exports, "CREATE_NATIVE_MINT_DISCRIMINATOR", {
  enumerable: true,
  get: function () { return token2022.CREATE_NATIVE_MINT_DISCRIMINATOR; }
});
Object.defineProperty(exports, "DISABLE_CONFIDENTIAL_CREDITS_CONFIDENTIAL_TRANSFER_DISCRIMINATOR", {
  enumerable: true,
  get: function () { return token2022.DISABLE_CONFIDENTIAL_CREDITS_CONFIDENTIAL_TRANSFER_DISCRIMINATOR; }
});
Object.defineProperty(exports, "DISABLE_CONFIDENTIAL_CREDITS_DISCRIMINATOR", {
  enumerable: true,
  get: function () { return token2022.DISABLE_CONFIDENTIAL_CREDITS_DISCRIMINATOR; }
});
Object.defineProperty(exports, "DISABLE_CPI_GUARD_CPI_GUARD_DISCRIMINATOR", {
  enumerable: true,
  get: function () { return token2022.DISABLE_CPI_GUARD_CPI_GUARD_DISCRIMINATOR; }
});
Object.defineProperty(exports, "DISABLE_CPI_GUARD_DISCRIMINATOR", {
  enumerable: true,
  get: function () { return token2022.DISABLE_CPI_GUARD_DISCRIMINATOR; }
});
Object.defineProperty(exports, "DISABLE_HARVEST_TO_MINT_CONFIDENTIAL_TRANSFER_FEE_DISCRIMINATOR", {
  enumerable: true,
  get: function () { return token2022.DISABLE_HARVEST_TO_MINT_CONFIDENTIAL_TRANSFER_FEE_DISCRIMINATOR; }
});
Object.defineProperty(exports, "DISABLE_HARVEST_TO_MINT_DISCRIMINATOR", {
  enumerable: true,
  get: function () { return token2022.DISABLE_HARVEST_TO_MINT_DISCRIMINATOR; }
});
Object.defineProperty(exports, "DISABLE_MEMO_TRANSFERS_DISCRIMINATOR", {
  enumerable: true,
  get: function () { return token2022.DISABLE_MEMO_TRANSFERS_DISCRIMINATOR; }
});
Object.defineProperty(exports, "DISABLE_MEMO_TRANSFERS_MEMO_TRANSFERS_DISCRIMINATOR", {
  enumerable: true,
  get: function () { return token2022.DISABLE_MEMO_TRANSFERS_MEMO_TRANSFERS_DISCRIMINATOR; }
});
Object.defineProperty(exports, "DISABLE_NON_CONFIDENTIAL_CREDITS_CONFIDENTIAL_TRANSFER_DISCRIMINATOR", {
  enumerable: true,
  get: function () { return token2022.DISABLE_NON_CONFIDENTIAL_CREDITS_CONFIDENTIAL_TRANSFER_DISCRIMINATOR; }
});
Object.defineProperty(exports, "DISABLE_NON_CONFIDENTIAL_CREDITS_DISCRIMINATOR", {
  enumerable: true,
  get: function () { return token2022.DISABLE_NON_CONFIDENTIAL_CREDITS_DISCRIMINATOR; }
});
Object.defineProperty(exports, "EMIT_TOKEN_METADATA_DISCRIMINATOR", {
  enumerable: true,
  get: function () { return token2022.EMIT_TOKEN_METADATA_DISCRIMINATOR; }
});
Object.defineProperty(exports, "EMPTY_CONFIDENTIAL_TRANSFER_ACCOUNT_CONFIDENTIAL_TRANSFER_DISCRIMINATOR", {
  enumerable: true,
  get: function () { return token2022.EMPTY_CONFIDENTIAL_TRANSFER_ACCOUNT_CONFIDENTIAL_TRANSFER_DISCRIMINATOR; }
});
Object.defineProperty(exports, "EMPTY_CONFIDENTIAL_TRANSFER_ACCOUNT_DISCRIMINATOR", {
  enumerable: true,
  get: function () { return token2022.EMPTY_CONFIDENTIAL_TRANSFER_ACCOUNT_DISCRIMINATOR; }
});
Object.defineProperty(exports, "ENABLE_CONFIDENTIAL_CREDITS_CONFIDENTIAL_TRANSFER_DISCRIMINATOR", {
  enumerable: true,
  get: function () { return token2022.ENABLE_CONFIDENTIAL_CREDITS_CONFIDENTIAL_TRANSFER_DISCRIMINATOR; }
});
Object.defineProperty(exports, "ENABLE_CONFIDENTIAL_CREDITS_DISCRIMINATOR", {
  enumerable: true,
  get: function () { return token2022.ENABLE_CONFIDENTIAL_CREDITS_DISCRIMINATOR; }
});
Object.defineProperty(exports, "ENABLE_CPI_GUARD_CPI_GUARD_DISCRIMINATOR", {
  enumerable: true,
  get: function () { return token2022.ENABLE_CPI_GUARD_CPI_GUARD_DISCRIMINATOR; }
});
Object.defineProperty(exports, "ENABLE_CPI_GUARD_DISCRIMINATOR", {
  enumerable: true,
  get: function () { return token2022.ENABLE_CPI_GUARD_DISCRIMINATOR; }
});
Object.defineProperty(exports, "ENABLE_HARVEST_TO_MINT_CONFIDENTIAL_TRANSFER_FEE_DISCRIMINATOR", {
  enumerable: true,
  get: function () { return token2022.ENABLE_HARVEST_TO_MINT_CONFIDENTIAL_TRANSFER_FEE_DISCRIMINATOR; }
});
Object.defineProperty(exports, "ENABLE_HARVEST_TO_MINT_DISCRIMINATOR", {
  enumerable: true,
  get: function () { return token2022.ENABLE_HARVEST_TO_MINT_DISCRIMINATOR; }
});
Object.defineProperty(exports, "ENABLE_MEMO_TRANSFERS_DISCRIMINATOR", {
  enumerable: true,
  get: function () { return token2022.ENABLE_MEMO_TRANSFERS_DISCRIMINATOR; }
});
Object.defineProperty(exports, "ENABLE_MEMO_TRANSFERS_MEMO_TRANSFERS_DISCRIMINATOR", {
  enumerable: true,
  get: function () { return token2022.ENABLE_MEMO_TRANSFERS_MEMO_TRANSFERS_DISCRIMINATOR; }
});
Object.defineProperty(exports, "ENABLE_NON_CONFIDENTIAL_CREDITS_CONFIDENTIAL_TRANSFER_DISCRIMINATOR", {
  enumerable: true,
  get: function () { return token2022.ENABLE_NON_CONFIDENTIAL_CREDITS_CONFIDENTIAL_TRANSFER_DISCRIMINATOR; }
});
Object.defineProperty(exports, "ENABLE_NON_CONFIDENTIAL_CREDITS_DISCRIMINATOR", {
  enumerable: true,
  get: function () { return token2022.ENABLE_NON_CONFIDENTIAL_CREDITS_DISCRIMINATOR; }
});
Object.defineProperty(exports, "ExtensionType", {
  enumerable: true,
  get: function () { return token2022.ExtensionType; }
});
Object.defineProperty(exports, "FREEZE_ACCOUNT_DISCRIMINATOR", {
  enumerable: true,
  get: function () { return token2022.FREEZE_ACCOUNT_DISCRIMINATOR; }
});
Object.defineProperty(exports, "GET_ACCOUNT_DATA_SIZE_DISCRIMINATOR", {
  enumerable: true,
  get: function () { return token2022.GET_ACCOUNT_DATA_SIZE_DISCRIMINATOR; }
});
Object.defineProperty(exports, "HARVEST_WITHHELD_TOKENS_TO_MINT_DISCRIMINATOR", {
  enumerable: true,
  get: function () { return token2022.HARVEST_WITHHELD_TOKENS_TO_MINT_DISCRIMINATOR; }
});
Object.defineProperty(exports, "HARVEST_WITHHELD_TOKENS_TO_MINT_FOR_CONFIDENTIAL_TRANSFER_FEE_CONFIDENTIAL_TRANSFER_FEE_DISCRIMINATOR", {
  enumerable: true,
  get: function () { return token2022.HARVEST_WITHHELD_TOKENS_TO_MINT_FOR_CONFIDENTIAL_TRANSFER_FEE_CONFIDENTIAL_TRANSFER_FEE_DISCRIMINATOR; }
});
Object.defineProperty(exports, "HARVEST_WITHHELD_TOKENS_TO_MINT_FOR_CONFIDENTIAL_TRANSFER_FEE_DISCRIMINATOR", {
  enumerable: true,
  get: function () { return token2022.HARVEST_WITHHELD_TOKENS_TO_MINT_FOR_CONFIDENTIAL_TRANSFER_FEE_DISCRIMINATOR; }
});
Object.defineProperty(exports, "HARVEST_WITHHELD_TOKENS_TO_MINT_TRANSFER_FEE_DISCRIMINATOR", {
  enumerable: true,
  get: function () { return token2022.HARVEST_WITHHELD_TOKENS_TO_MINT_TRANSFER_FEE_DISCRIMINATOR; }
});
Object.defineProperty(exports, "INITIALIZE_ACCOUNT2_DISCRIMINATOR", {
  enumerable: true,
  get: function () { return token2022.INITIALIZE_ACCOUNT2_DISCRIMINATOR; }
});
Object.defineProperty(exports, "INITIALIZE_ACCOUNT3_DISCRIMINATOR", {
  enumerable: true,
  get: function () { return token2022.INITIALIZE_ACCOUNT3_DISCRIMINATOR; }
});
Object.defineProperty(exports, "INITIALIZE_ACCOUNT_DISCRIMINATOR", {
  enumerable: true,
  get: function () { return token2022.INITIALIZE_ACCOUNT_DISCRIMINATOR; }
});
Object.defineProperty(exports, "INITIALIZE_CONFIDENTIAL_TRANSFER_FEE_CONFIDENTIAL_TRANSFER_FEE_DISCRIMINATOR", {
  enumerable: true,
  get: function () { return token2022.INITIALIZE_CONFIDENTIAL_TRANSFER_FEE_CONFIDENTIAL_TRANSFER_FEE_DISCRIMINATOR; }
});
Object.defineProperty(exports, "INITIALIZE_CONFIDENTIAL_TRANSFER_FEE_DISCRIMINATOR", {
  enumerable: true,
  get: function () { return token2022.INITIALIZE_CONFIDENTIAL_TRANSFER_FEE_DISCRIMINATOR; }
});
Object.defineProperty(exports, "INITIALIZE_CONFIDENTIAL_TRANSFER_MINT_CONFIDENTIAL_TRANSFER_DISCRIMINATOR", {
  enumerable: true,
  get: function () { return token2022.INITIALIZE_CONFIDENTIAL_TRANSFER_MINT_CONFIDENTIAL_TRANSFER_DISCRIMINATOR; }
});
Object.defineProperty(exports, "INITIALIZE_CONFIDENTIAL_TRANSFER_MINT_DISCRIMINATOR", {
  enumerable: true,
  get: function () { return token2022.INITIALIZE_CONFIDENTIAL_TRANSFER_MINT_DISCRIMINATOR; }
});
Object.defineProperty(exports, "INITIALIZE_DEFAULT_ACCOUNT_STATE_DEFAULT_ACCOUNT_STATE_DISCRIMINATOR", {
  enumerable: true,
  get: function () { return token2022.INITIALIZE_DEFAULT_ACCOUNT_STATE_DEFAULT_ACCOUNT_STATE_DISCRIMINATOR; }
});
Object.defineProperty(exports, "INITIALIZE_DEFAULT_ACCOUNT_STATE_DISCRIMINATOR", {
  enumerable: true,
  get: function () { return token2022.INITIALIZE_DEFAULT_ACCOUNT_STATE_DISCRIMINATOR; }
});
Object.defineProperty(exports, "INITIALIZE_GROUP_MEMBER_POINTER_DISCRIMINATOR", {
  enumerable: true,
  get: function () { return token2022.INITIALIZE_GROUP_MEMBER_POINTER_DISCRIMINATOR; }
});
Object.defineProperty(exports, "INITIALIZE_GROUP_MEMBER_POINTER_GROUP_MEMBER_POINTER_DISCRIMINATOR", {
  enumerable: true,
  get: function () { return token2022.INITIALIZE_GROUP_MEMBER_POINTER_GROUP_MEMBER_POINTER_DISCRIMINATOR; }
});
Object.defineProperty(exports, "INITIALIZE_GROUP_POINTER_DISCRIMINATOR", {
  enumerable: true,
  get: function () { return token2022.INITIALIZE_GROUP_POINTER_DISCRIMINATOR; }
});
Object.defineProperty(exports, "INITIALIZE_GROUP_POINTER_GROUP_POINTER_DISCRIMINATOR", {
  enumerable: true,
  get: function () { return token2022.INITIALIZE_GROUP_POINTER_GROUP_POINTER_DISCRIMINATOR; }
});
Object.defineProperty(exports, "INITIALIZE_IMMUTABLE_OWNER_DISCRIMINATOR", {
  enumerable: true,
  get: function () { return token2022.INITIALIZE_IMMUTABLE_OWNER_DISCRIMINATOR; }
});
Object.defineProperty(exports, "INITIALIZE_INTEREST_BEARING_MINT_DISCRIMINATOR", {
  enumerable: true,
  get: function () { return token2022.INITIALIZE_INTEREST_BEARING_MINT_DISCRIMINATOR; }
});
Object.defineProperty(exports, "INITIALIZE_INTEREST_BEARING_MINT_INTEREST_BEARING_MINT_DISCRIMINATOR", {
  enumerable: true,
  get: function () { return token2022.INITIALIZE_INTEREST_BEARING_MINT_INTEREST_BEARING_MINT_DISCRIMINATOR; }
});
Object.defineProperty(exports, "INITIALIZE_METADATA_POINTER_DISCRIMINATOR", {
  enumerable: true,
  get: function () { return token2022.INITIALIZE_METADATA_POINTER_DISCRIMINATOR; }
});
Object.defineProperty(exports, "INITIALIZE_METADATA_POINTER_METADATA_POINTER_DISCRIMINATOR", {
  enumerable: true,
  get: function () { return token2022.INITIALIZE_METADATA_POINTER_METADATA_POINTER_DISCRIMINATOR; }
});
Object.defineProperty(exports, "INITIALIZE_MINT2_DISCRIMINATOR", {
  enumerable: true,
  get: function () { return token2022.INITIALIZE_MINT2_DISCRIMINATOR; }
});
Object.defineProperty(exports, "INITIALIZE_MINT_CLOSE_AUTHORITY_DISCRIMINATOR", {
  enumerable: true,
  get: function () { return token2022.INITIALIZE_MINT_CLOSE_AUTHORITY_DISCRIMINATOR; }
});
Object.defineProperty(exports, "INITIALIZE_MINT_DISCRIMINATOR", {
  enumerable: true,
  get: function () { return token2022.INITIALIZE_MINT_DISCRIMINATOR; }
});
Object.defineProperty(exports, "INITIALIZE_MULTISIG2_DISCRIMINATOR", {
  enumerable: true,
  get: function () { return token2022.INITIALIZE_MULTISIG2_DISCRIMINATOR; }
});
Object.defineProperty(exports, "INITIALIZE_MULTISIG_DISCRIMINATOR", {
  enumerable: true,
  get: function () { return token2022.INITIALIZE_MULTISIG_DISCRIMINATOR; }
});
Object.defineProperty(exports, "INITIALIZE_NON_TRANSFERABLE_MINT_DISCRIMINATOR", {
  enumerable: true,
  get: function () { return token2022.INITIALIZE_NON_TRANSFERABLE_MINT_DISCRIMINATOR; }
});
Object.defineProperty(exports, "INITIALIZE_PAUSABLE_CONFIG_DISCRIMINATOR", {
  enumerable: true,
  get: function () { return token2022.INITIALIZE_PAUSABLE_CONFIG_DISCRIMINATOR; }
});
Object.defineProperty(exports, "INITIALIZE_PAUSABLE_CONFIG_PAUSABLE_DISCRIMINATOR", {
  enumerable: true,
  get: function () { return token2022.INITIALIZE_PAUSABLE_CONFIG_PAUSABLE_DISCRIMINATOR; }
});
Object.defineProperty(exports, "INITIALIZE_PERMANENT_DELEGATE_DISCRIMINATOR", {
  enumerable: true,
  get: function () { return token2022.INITIALIZE_PERMANENT_DELEGATE_DISCRIMINATOR; }
});
Object.defineProperty(exports, "INITIALIZE_SCALED_UI_AMOUNT_MINT_DISCRIMINATOR", {
  enumerable: true,
  get: function () { return token2022.INITIALIZE_SCALED_UI_AMOUNT_MINT_DISCRIMINATOR; }
});
Object.defineProperty(exports, "INITIALIZE_SCALED_UI_AMOUNT_MINT_SCALED_UI_AMOUNT_MINT_DISCRIMINATOR", {
  enumerable: true,
  get: function () { return token2022.INITIALIZE_SCALED_UI_AMOUNT_MINT_SCALED_UI_AMOUNT_MINT_DISCRIMINATOR; }
});
Object.defineProperty(exports, "INITIALIZE_TOKEN_GROUP_DISCRIMINATOR", {
  enumerable: true,
  get: function () { return token2022.INITIALIZE_TOKEN_GROUP_DISCRIMINATOR; }
});
Object.defineProperty(exports, "INITIALIZE_TOKEN_GROUP_MEMBER_DISCRIMINATOR", {
  enumerable: true,
  get: function () { return token2022.INITIALIZE_TOKEN_GROUP_MEMBER_DISCRIMINATOR; }
});
Object.defineProperty(exports, "INITIALIZE_TOKEN_METADATA_DISCRIMINATOR", {
  enumerable: true,
  get: function () { return token2022.INITIALIZE_TOKEN_METADATA_DISCRIMINATOR; }
});
Object.defineProperty(exports, "INITIALIZE_TRANSFER_FEE_CONFIG_DISCRIMINATOR", {
  enumerable: true,
  get: function () { return token2022.INITIALIZE_TRANSFER_FEE_CONFIG_DISCRIMINATOR; }
});
Object.defineProperty(exports, "INITIALIZE_TRANSFER_FEE_CONFIG_TRANSFER_FEE_DISCRIMINATOR", {
  enumerable: true,
  get: function () { return token2022.INITIALIZE_TRANSFER_FEE_CONFIG_TRANSFER_FEE_DISCRIMINATOR; }
});
Object.defineProperty(exports, "INITIALIZE_TRANSFER_HOOK_DISCRIMINATOR", {
  enumerable: true,
  get: function () { return token2022.INITIALIZE_TRANSFER_HOOK_DISCRIMINATOR; }
});
Object.defineProperty(exports, "INITIALIZE_TRANSFER_HOOK_TRANSFER_HOOK_DISCRIMINATOR", {
  enumerable: true,
  get: function () { return token2022.INITIALIZE_TRANSFER_HOOK_TRANSFER_HOOK_DISCRIMINATOR; }
});
Object.defineProperty(exports, "MINT_TO_CHECKED_DISCRIMINATOR", {
  enumerable: true,
  get: function () { return token2022.MINT_TO_CHECKED_DISCRIMINATOR; }
});
Object.defineProperty(exports, "MINT_TO_DISCRIMINATOR", {
  enumerable: true,
  get: function () { return token2022.MINT_TO_DISCRIMINATOR; }
});
Object.defineProperty(exports, "PAUSE_DISCRIMINATOR", {
  enumerable: true,
  get: function () { return token2022.PAUSE_DISCRIMINATOR; }
});
Object.defineProperty(exports, "PAUSE_PAUSABLE_DISCRIMINATOR", {
  enumerable: true,
  get: function () { return token2022.PAUSE_PAUSABLE_DISCRIMINATOR; }
});
Object.defineProperty(exports, "REALLOCATE_DISCRIMINATOR", {
  enumerable: true,
  get: function () { return token2022.REALLOCATE_DISCRIMINATOR; }
});
Object.defineProperty(exports, "RECOVER_NESTED_ASSOCIATED_TOKEN_DISCRIMINATOR", {
  enumerable: true,
  get: function () { return token2022.RECOVER_NESTED_ASSOCIATED_TOKEN_DISCRIMINATOR; }
});
Object.defineProperty(exports, "REMOVE_TOKEN_METADATA_KEY_DISCRIMINATOR", {
  enumerable: true,
  get: function () { return token2022.REMOVE_TOKEN_METADATA_KEY_DISCRIMINATOR; }
});
Object.defineProperty(exports, "RESUME_DISCRIMINATOR", {
  enumerable: true,
  get: function () { return token2022.RESUME_DISCRIMINATOR; }
});
Object.defineProperty(exports, "RESUME_PAUSABLE_DISCRIMINATOR", {
  enumerable: true,
  get: function () { return token2022.RESUME_PAUSABLE_DISCRIMINATOR; }
});
Object.defineProperty(exports, "REVOKE_DISCRIMINATOR", {
  enumerable: true,
  get: function () { return token2022.REVOKE_DISCRIMINATOR; }
});
Object.defineProperty(exports, "SET_AUTHORITY_DISCRIMINATOR", {
  enumerable: true,
  get: function () { return token2022.SET_AUTHORITY_DISCRIMINATOR; }
});
Object.defineProperty(exports, "SET_TRANSFER_FEE_DISCRIMINATOR", {
  enumerable: true,
  get: function () { return token2022.SET_TRANSFER_FEE_DISCRIMINATOR; }
});
Object.defineProperty(exports, "SET_TRANSFER_FEE_TRANSFER_FEE_DISCRIMINATOR", {
  enumerable: true,
  get: function () { return token2022.SET_TRANSFER_FEE_TRANSFER_FEE_DISCRIMINATOR; }
});
Object.defineProperty(exports, "SYNC_NATIVE_DISCRIMINATOR", {
  enumerable: true,
  get: function () { return token2022.SYNC_NATIVE_DISCRIMINATOR; }
});
Object.defineProperty(exports, "THAW_ACCOUNT_DISCRIMINATOR", {
  enumerable: true,
  get: function () { return token2022.THAW_ACCOUNT_DISCRIMINATOR; }
});
Object.defineProperty(exports, "TOKEN_2022_ERROR__ACCOUNT_FROZEN", {
  enumerable: true,
  get: function () { return token2022.TOKEN_2022_ERROR__ACCOUNT_FROZEN; }
});
Object.defineProperty(exports, "TOKEN_2022_ERROR__ALREADY_IN_USE", {
  enumerable: true,
  get: function () { return token2022.TOKEN_2022_ERROR__ALREADY_IN_USE; }
});
Object.defineProperty(exports, "TOKEN_2022_ERROR__AUTHORITY_TYPE_NOT_SUPPORTED", {
  enumerable: true,
  get: function () { return token2022.TOKEN_2022_ERROR__AUTHORITY_TYPE_NOT_SUPPORTED; }
});
Object.defineProperty(exports, "TOKEN_2022_ERROR__FIXED_SUPPLY", {
  enumerable: true,
  get: function () { return token2022.TOKEN_2022_ERROR__FIXED_SUPPLY; }
});
Object.defineProperty(exports, "TOKEN_2022_ERROR__INSUFFICIENT_FUNDS", {
  enumerable: true,
  get: function () { return token2022.TOKEN_2022_ERROR__INSUFFICIENT_FUNDS; }
});
Object.defineProperty(exports, "TOKEN_2022_ERROR__INVALID_INSTRUCTION", {
  enumerable: true,
  get: function () { return token2022.TOKEN_2022_ERROR__INVALID_INSTRUCTION; }
});
Object.defineProperty(exports, "TOKEN_2022_ERROR__INVALID_MINT", {
  enumerable: true,
  get: function () { return token2022.TOKEN_2022_ERROR__INVALID_MINT; }
});
Object.defineProperty(exports, "TOKEN_2022_ERROR__INVALID_NUMBER_OF_PROVIDED_SIGNERS", {
  enumerable: true,
  get: function () { return token2022.TOKEN_2022_ERROR__INVALID_NUMBER_OF_PROVIDED_SIGNERS; }
});
Object.defineProperty(exports, "TOKEN_2022_ERROR__INVALID_NUMBER_OF_REQUIRED_SIGNERS", {
  enumerable: true,
  get: function () { return token2022.TOKEN_2022_ERROR__INVALID_NUMBER_OF_REQUIRED_SIGNERS; }
});
Object.defineProperty(exports, "TOKEN_2022_ERROR__INVALID_STATE", {
  enumerable: true,
  get: function () { return token2022.TOKEN_2022_ERROR__INVALID_STATE; }
});
Object.defineProperty(exports, "TOKEN_2022_ERROR__MINT_CANNOT_FREEZE", {
  enumerable: true,
  get: function () { return token2022.TOKEN_2022_ERROR__MINT_CANNOT_FREEZE; }
});
Object.defineProperty(exports, "TOKEN_2022_ERROR__MINT_DECIMALS_MISMATCH", {
  enumerable: true,
  get: function () { return token2022.TOKEN_2022_ERROR__MINT_DECIMALS_MISMATCH; }
});
Object.defineProperty(exports, "TOKEN_2022_ERROR__MINT_MISMATCH", {
  enumerable: true,
  get: function () { return token2022.TOKEN_2022_ERROR__MINT_MISMATCH; }
});
Object.defineProperty(exports, "TOKEN_2022_ERROR__NATIVE_NOT_SUPPORTED", {
  enumerable: true,
  get: function () { return token2022.TOKEN_2022_ERROR__NATIVE_NOT_SUPPORTED; }
});
Object.defineProperty(exports, "TOKEN_2022_ERROR__NON_NATIVE_HAS_BALANCE", {
  enumerable: true,
  get: function () { return token2022.TOKEN_2022_ERROR__NON_NATIVE_HAS_BALANCE; }
});
Object.defineProperty(exports, "TOKEN_2022_ERROR__NON_NATIVE_NOT_SUPPORTED", {
  enumerable: true,
  get: function () { return token2022.TOKEN_2022_ERROR__NON_NATIVE_NOT_SUPPORTED; }
});
Object.defineProperty(exports, "TOKEN_2022_ERROR__NOT_RENT_EXEMPT", {
  enumerable: true,
  get: function () { return token2022.TOKEN_2022_ERROR__NOT_RENT_EXEMPT; }
});
Object.defineProperty(exports, "TOKEN_2022_ERROR__OVERFLOW", {
  enumerable: true,
  get: function () { return token2022.TOKEN_2022_ERROR__OVERFLOW; }
});
Object.defineProperty(exports, "TOKEN_2022_ERROR__OWNER_MISMATCH", {
  enumerable: true,
  get: function () { return token2022.TOKEN_2022_ERROR__OWNER_MISMATCH; }
});
Object.defineProperty(exports, "TOKEN_2022_ERROR__UNINITIALIZED_STATE", {
  enumerable: true,
  get: function () { return token2022.TOKEN_2022_ERROR__UNINITIALIZED_STATE; }
});
Object.defineProperty(exports, "TOKEN_2022_PROGRAM_ADDRESS", {
  enumerable: true,
  get: function () { return token2022.TOKEN_2022_PROGRAM_ADDRESS; }
});
Object.defineProperty(exports, "TRANSFER_CHECKED_DISCRIMINATOR", {
  enumerable: true,
  get: function () { return token2022.TRANSFER_CHECKED_DISCRIMINATOR; }
});
Object.defineProperty(exports, "TRANSFER_CHECKED_WITH_FEE_DISCRIMINATOR", {
  enumerable: true,
  get: function () { return token2022.TRANSFER_CHECKED_WITH_FEE_DISCRIMINATOR; }
});
Object.defineProperty(exports, "TRANSFER_CHECKED_WITH_FEE_TRANSFER_FEE_DISCRIMINATOR", {
  enumerable: true,
  get: function () { return token2022.TRANSFER_CHECKED_WITH_FEE_TRANSFER_FEE_DISCRIMINATOR; }
});
Object.defineProperty(exports, "TRANSFER_DISCRIMINATOR", {
  enumerable: true,
  get: function () { return token2022.TRANSFER_DISCRIMINATOR; }
});
Object.defineProperty(exports, "Token2022Account", {
  enumerable: true,
  get: function () { return token2022.Token2022Account; }
});
Object.defineProperty(exports, "Token2022Instruction", {
  enumerable: true,
  get: function () { return token2022.Token2022Instruction; }
});
Object.defineProperty(exports, "UI_AMOUNT_TO_AMOUNT_DISCRIMINATOR", {
  enumerable: true,
  get: function () { return token2022.UI_AMOUNT_TO_AMOUNT_DISCRIMINATOR; }
});
Object.defineProperty(exports, "UPDATE_CONFIDENTIAL_TRANSFER_MINT_CONFIDENTIAL_TRANSFER_DISCRIMINATOR", {
  enumerable: true,
  get: function () { return token2022.UPDATE_CONFIDENTIAL_TRANSFER_MINT_CONFIDENTIAL_TRANSFER_DISCRIMINATOR; }
});
Object.defineProperty(exports, "UPDATE_CONFIDENTIAL_TRANSFER_MINT_DISCRIMINATOR", {
  enumerable: true,
  get: function () { return token2022.UPDATE_CONFIDENTIAL_TRANSFER_MINT_DISCRIMINATOR; }
});
Object.defineProperty(exports, "UPDATE_DEFAULT_ACCOUNT_STATE_DEFAULT_ACCOUNT_STATE_DISCRIMINATOR", {
  enumerable: true,
  get: function () { return token2022.UPDATE_DEFAULT_ACCOUNT_STATE_DEFAULT_ACCOUNT_STATE_DISCRIMINATOR; }
});
Object.defineProperty(exports, "UPDATE_DEFAULT_ACCOUNT_STATE_DISCRIMINATOR", {
  enumerable: true,
  get: function () { return token2022.UPDATE_DEFAULT_ACCOUNT_STATE_DISCRIMINATOR; }
});
Object.defineProperty(exports, "UPDATE_GROUP_MEMBER_POINTER_DISCRIMINATOR", {
  enumerable: true,
  get: function () { return token2022.UPDATE_GROUP_MEMBER_POINTER_DISCRIMINATOR; }
});
Object.defineProperty(exports, "UPDATE_GROUP_MEMBER_POINTER_GROUP_MEMBER_POINTER_DISCRIMINATOR", {
  enumerable: true,
  get: function () { return token2022.UPDATE_GROUP_MEMBER_POINTER_GROUP_MEMBER_POINTER_DISCRIMINATOR; }
});
Object.defineProperty(exports, "UPDATE_GROUP_POINTER_DISCRIMINATOR", {
  enumerable: true,
  get: function () { return token2022.UPDATE_GROUP_POINTER_DISCRIMINATOR; }
});
Object.defineProperty(exports, "UPDATE_GROUP_POINTER_GROUP_POINTER_DISCRIMINATOR", {
  enumerable: true,
  get: function () { return token2022.UPDATE_GROUP_POINTER_GROUP_POINTER_DISCRIMINATOR; }
});
Object.defineProperty(exports, "UPDATE_METADATA_POINTER_DISCRIMINATOR", {
  enumerable: true,
  get: function () { return token2022.UPDATE_METADATA_POINTER_DISCRIMINATOR; }
});
Object.defineProperty(exports, "UPDATE_METADATA_POINTER_METADATA_POINTER_DISCRIMINATOR", {
  enumerable: true,
  get: function () { return token2022.UPDATE_METADATA_POINTER_METADATA_POINTER_DISCRIMINATOR; }
});
Object.defineProperty(exports, "UPDATE_MULTIPLIER_SCALED_UI_MINT_DISCRIMINATOR", {
  enumerable: true,
  get: function () { return token2022.UPDATE_MULTIPLIER_SCALED_UI_MINT_DISCRIMINATOR; }
});
Object.defineProperty(exports, "UPDATE_MULTIPLIER_SCALED_UI_MINT_SCALED_UI_AMOUNT_MINT_DISCRIMINATOR", {
  enumerable: true,
  get: function () { return token2022.UPDATE_MULTIPLIER_SCALED_UI_MINT_SCALED_UI_AMOUNT_MINT_DISCRIMINATOR; }
});
Object.defineProperty(exports, "UPDATE_RATE_INTEREST_BEARING_MINT_DISCRIMINATOR", {
  enumerable: true,
  get: function () { return token2022.UPDATE_RATE_INTEREST_BEARING_MINT_DISCRIMINATOR; }
});
Object.defineProperty(exports, "UPDATE_RATE_INTEREST_BEARING_MINT_INTEREST_BEARING_MINT_DISCRIMINATOR", {
  enumerable: true,
  get: function () { return token2022.UPDATE_RATE_INTEREST_BEARING_MINT_INTEREST_BEARING_MINT_DISCRIMINATOR; }
});
Object.defineProperty(exports, "UPDATE_TOKEN_GROUP_MAX_SIZE_DISCRIMINATOR", {
  enumerable: true,
  get: function () { return token2022.UPDATE_TOKEN_GROUP_MAX_SIZE_DISCRIMINATOR; }
});
Object.defineProperty(exports, "UPDATE_TOKEN_GROUP_UPDATE_AUTHORITY_DISCRIMINATOR", {
  enumerable: true,
  get: function () { return token2022.UPDATE_TOKEN_GROUP_UPDATE_AUTHORITY_DISCRIMINATOR; }
});
Object.defineProperty(exports, "UPDATE_TOKEN_METADATA_FIELD_DISCRIMINATOR", {
  enumerable: true,
  get: function () { return token2022.UPDATE_TOKEN_METADATA_FIELD_DISCRIMINATOR; }
});
Object.defineProperty(exports, "UPDATE_TOKEN_METADATA_UPDATE_AUTHORITY_DISCRIMINATOR", {
  enumerable: true,
  get: function () { return token2022.UPDATE_TOKEN_METADATA_UPDATE_AUTHORITY_DISCRIMINATOR; }
});
Object.defineProperty(exports, "UPDATE_TRANSFER_HOOK_DISCRIMINATOR", {
  enumerable: true,
  get: function () { return token2022.UPDATE_TRANSFER_HOOK_DISCRIMINATOR; }
});
Object.defineProperty(exports, "UPDATE_TRANSFER_HOOK_TRANSFER_HOOK_DISCRIMINATOR", {
  enumerable: true,
  get: function () { return token2022.UPDATE_TRANSFER_HOOK_TRANSFER_HOOK_DISCRIMINATOR; }
});
Object.defineProperty(exports, "WITHDRAW_EXCESS_LAMPORTS_DISCRIMINATOR", {
  enumerable: true,
  get: function () { return token2022.WITHDRAW_EXCESS_LAMPORTS_DISCRIMINATOR; }
});
Object.defineProperty(exports, "WITHDRAW_WITHHELD_TOKENS_FROM_ACCOUNTS_DISCRIMINATOR", {
  enumerable: true,
  get: function () { return token2022.WITHDRAW_WITHHELD_TOKENS_FROM_ACCOUNTS_DISCRIMINATOR; }
});
Object.defineProperty(exports, "WITHDRAW_WITHHELD_TOKENS_FROM_ACCOUNTS_FOR_CONFIDENTIAL_TRANSFER_FEE_CONFIDENTIAL_TRANSFER_FEE_DISCRIMINATOR", {
  enumerable: true,
  get: function () { return token2022.WITHDRAW_WITHHELD_TOKENS_FROM_ACCOUNTS_FOR_CONFIDENTIAL_TRANSFER_FEE_CONFIDENTIAL_TRANSFER_FEE_DISCRIMINATOR; }
});
Object.defineProperty(exports, "WITHDRAW_WITHHELD_TOKENS_FROM_ACCOUNTS_FOR_CONFIDENTIAL_TRANSFER_FEE_DISCRIMINATOR", {
  enumerable: true,
  get: function () { return token2022.WITHDRAW_WITHHELD_TOKENS_FROM_ACCOUNTS_FOR_CONFIDENTIAL_TRANSFER_FEE_DISCRIMINATOR; }
});
Object.defineProperty(exports, "WITHDRAW_WITHHELD_TOKENS_FROM_ACCOUNTS_TRANSFER_FEE_DISCRIMINATOR", {
  enumerable: true,
  get: function () { return token2022.WITHDRAW_WITHHELD_TOKENS_FROM_ACCOUNTS_TRANSFER_FEE_DISCRIMINATOR; }
});
Object.defineProperty(exports, "WITHDRAW_WITHHELD_TOKENS_FROM_MINT_DISCRIMINATOR", {
  enumerable: true,
  get: function () { return token2022.WITHDRAW_WITHHELD_TOKENS_FROM_MINT_DISCRIMINATOR; }
});
Object.defineProperty(exports, "WITHDRAW_WITHHELD_TOKENS_FROM_MINT_FOR_CONFIDENTIAL_TRANSFER_FEE_CONFIDENTIAL_TRANSFER_FEE_DISCRIMINATOR", {
  enumerable: true,
  get: function () { return token2022.WITHDRAW_WITHHELD_TOKENS_FROM_MINT_FOR_CONFIDENTIAL_TRANSFER_FEE_CONFIDENTIAL_TRANSFER_FEE_DISCRIMINATOR; }
});
Object.defineProperty(exports, "WITHDRAW_WITHHELD_TOKENS_FROM_MINT_FOR_CONFIDENTIAL_TRANSFER_FEE_DISCRIMINATOR", {
  enumerable: true,
  get: function () { return token2022.WITHDRAW_WITHHELD_TOKENS_FROM_MINT_FOR_CONFIDENTIAL_TRANSFER_FEE_DISCRIMINATOR; }
});
Object.defineProperty(exports, "WITHDRAW_WITHHELD_TOKENS_FROM_MINT_TRANSFER_FEE_DISCRIMINATOR", {
  enumerable: true,
  get: function () { return token2022.WITHDRAW_WITHHELD_TOKENS_FROM_MINT_TRANSFER_FEE_DISCRIMINATOR; }
});
Object.defineProperty(exports, "amountToUiAmountForInterestBearingMintWithoutSimulation", {
  enumerable: true,
  get: function () { return token2022.amountToUiAmountForInterestBearingMintWithoutSimulation; }
});
Object.defineProperty(exports, "amountToUiAmountForMintWithoutSimulation", {
  enumerable: true,
  get: function () { return token2022.amountToUiAmountForMintWithoutSimulation; }
});
Object.defineProperty(exports, "amountToUiAmountForScaledUiAmountMintWithoutSimulation", {
  enumerable: true,
  get: function () { return token2022.amountToUiAmountForScaledUiAmountMintWithoutSimulation; }
});
Object.defineProperty(exports, "decodeMint", {
  enumerable: true,
  get: function () { return token2022.decodeMint; }
});
Object.defineProperty(exports, "decodeMultisig", {
  enumerable: true,
  get: function () { return token2022.decodeMultisig; }
});
Object.defineProperty(exports, "decodeToken", {
  enumerable: true,
  get: function () { return token2022.decodeToken; }
});
Object.defineProperty(exports, "extension", {
  enumerable: true,
  get: function () { return token2022.extension; }
});
Object.defineProperty(exports, "fetchAllMaybeMint", {
  enumerable: true,
  get: function () { return token2022.fetchAllMaybeMint; }
});
Object.defineProperty(exports, "fetchAllMaybeMultisig", {
  enumerable: true,
  get: function () { return token2022.fetchAllMaybeMultisig; }
});
Object.defineProperty(exports, "fetchAllMaybeToken", {
  enumerable: true,
  get: function () { return token2022.fetchAllMaybeToken; }
});
Object.defineProperty(exports, "fetchAllMint", {
  enumerable: true,
  get: function () { return token2022.fetchAllMint; }
});
Object.defineProperty(exports, "fetchAllMultisig", {
  enumerable: true,
  get: function () { return token2022.fetchAllMultisig; }
});
Object.defineProperty(exports, "fetchAllToken", {
  enumerable: true,
  get: function () { return token2022.fetchAllToken; }
});
Object.defineProperty(exports, "fetchMaybeMint", {
  enumerable: true,
  get: function () { return token2022.fetchMaybeMint; }
});
Object.defineProperty(exports, "fetchMaybeMultisig", {
  enumerable: true,
  get: function () { return token2022.fetchMaybeMultisig; }
});
Object.defineProperty(exports, "fetchMaybeToken", {
  enumerable: true,
  get: function () { return token2022.fetchMaybeToken; }
});
Object.defineProperty(exports, "fetchMint", {
  enumerable: true,
  get: function () { return token2022.fetchMint; }
});
Object.defineProperty(exports, "fetchMultisig", {
  enumerable: true,
  get: function () { return token2022.fetchMultisig; }
});
Object.defineProperty(exports, "fetchToken", {
  enumerable: true,
  get: function () { return token2022.fetchToken; }
});
Object.defineProperty(exports, "findAssociatedTokenPda", {
  enumerable: true,
  get: function () { return token2022.findAssociatedTokenPda; }
});
Object.defineProperty(exports, "getAccountStateCodec", {
  enumerable: true,
  get: function () { return token2022.getAccountStateCodec; }
});
Object.defineProperty(exports, "getAccountStateDecoder", {
  enumerable: true,
  get: function () { return token2022.getAccountStateDecoder; }
});
Object.defineProperty(exports, "getAccountStateEncoder", {
  enumerable: true,
  get: function () { return token2022.getAccountStateEncoder; }
});
Object.defineProperty(exports, "getAmountToUiAmountDiscriminatorBytes", {
  enumerable: true,
  get: function () { return token2022.getAmountToUiAmountDiscriminatorBytes; }
});
Object.defineProperty(exports, "getAmountToUiAmountInstruction", {
  enumerable: true,
  get: function () { return token2022.getAmountToUiAmountInstruction; }
});
Object.defineProperty(exports, "getAmountToUiAmountInstructionDataCodec", {
  enumerable: true,
  get: function () { return token2022.getAmountToUiAmountInstructionDataCodec; }
});
Object.defineProperty(exports, "getAmountToUiAmountInstructionDataDecoder", {
  enumerable: true,
  get: function () { return token2022.getAmountToUiAmountInstructionDataDecoder; }
});
Object.defineProperty(exports, "getAmountToUiAmountInstructionDataEncoder", {
  enumerable: true,
  get: function () { return token2022.getAmountToUiAmountInstructionDataEncoder; }
});
Object.defineProperty(exports, "getApplyConfidentialPendingBalanceConfidentialTransferDiscriminatorBytes", {
  enumerable: true,
  get: function () { return token2022.getApplyConfidentialPendingBalanceConfidentialTransferDiscriminatorBytes; }
});
Object.defineProperty(exports, "getApplyConfidentialPendingBalanceDiscriminatorBytes", {
  enumerable: true,
  get: function () { return token2022.getApplyConfidentialPendingBalanceDiscriminatorBytes; }
});
Object.defineProperty(exports, "getApplyConfidentialPendingBalanceInstruction", {
  enumerable: true,
  get: function () { return token2022.getApplyConfidentialPendingBalanceInstruction; }
});
Object.defineProperty(exports, "getApplyConfidentialPendingBalanceInstructionDataCodec", {
  enumerable: true,
  get: function () { return token2022.getApplyConfidentialPendingBalanceInstructionDataCodec; }
});
Object.defineProperty(exports, "getApplyConfidentialPendingBalanceInstructionDataDecoder", {
  enumerable: true,
  get: function () { return token2022.getApplyConfidentialPendingBalanceInstructionDataDecoder; }
});
Object.defineProperty(exports, "getApplyConfidentialPendingBalanceInstructionDataEncoder", {
  enumerable: true,
  get: function () { return token2022.getApplyConfidentialPendingBalanceInstructionDataEncoder; }
});
Object.defineProperty(exports, "getApproveCheckedDiscriminatorBytes", {
  enumerable: true,
  get: function () { return token2022.getApproveCheckedDiscriminatorBytes; }
});
Object.defineProperty(exports, "getApproveCheckedInstruction", {
  enumerable: true,
  get: function () { return token2022.getApproveCheckedInstruction; }
});
Object.defineProperty(exports, "getApproveCheckedInstructionDataCodec", {
  enumerable: true,
  get: function () { return token2022.getApproveCheckedInstructionDataCodec; }
});
Object.defineProperty(exports, "getApproveCheckedInstructionDataDecoder", {
  enumerable: true,
  get: function () { return token2022.getApproveCheckedInstructionDataDecoder; }
});
Object.defineProperty(exports, "getApproveCheckedInstructionDataEncoder", {
  enumerable: true,
  get: function () { return token2022.getApproveCheckedInstructionDataEncoder; }
});
Object.defineProperty(exports, "getApproveConfidentialTransferAccountConfidentialTransferDiscriminatorBytes", {
  enumerable: true,
  get: function () { return token2022.getApproveConfidentialTransferAccountConfidentialTransferDiscriminatorBytes; }
});
Object.defineProperty(exports, "getApproveConfidentialTransferAccountDiscriminatorBytes", {
  enumerable: true,
  get: function () { return token2022.getApproveConfidentialTransferAccountDiscriminatorBytes; }
});
Object.defineProperty(exports, "getApproveConfidentialTransferAccountInstruction", {
  enumerable: true,
  get: function () { return token2022.getApproveConfidentialTransferAccountInstruction; }
});
Object.defineProperty(exports, "getApproveConfidentialTransferAccountInstructionDataCodec", {
  enumerable: true,
  get: function () { return token2022.getApproveConfidentialTransferAccountInstructionDataCodec; }
});
Object.defineProperty(exports, "getApproveConfidentialTransferAccountInstructionDataDecoder", {
  enumerable: true,
  get: function () { return token2022.getApproveConfidentialTransferAccountInstructionDataDecoder; }
});
Object.defineProperty(exports, "getApproveConfidentialTransferAccountInstructionDataEncoder", {
  enumerable: true,
  get: function () { return token2022.getApproveConfidentialTransferAccountInstructionDataEncoder; }
});
Object.defineProperty(exports, "getApproveDiscriminatorBytes", {
  enumerable: true,
  get: function () { return token2022.getApproveDiscriminatorBytes; }
});
Object.defineProperty(exports, "getApproveInstruction", {
  enumerable: true,
  get: function () { return token2022.getApproveInstruction; }
});
Object.defineProperty(exports, "getApproveInstructionDataCodec", {
  enumerable: true,
  get: function () { return token2022.getApproveInstructionDataCodec; }
});
Object.defineProperty(exports, "getApproveInstructionDataDecoder", {
  enumerable: true,
  get: function () { return token2022.getApproveInstructionDataDecoder; }
});
Object.defineProperty(exports, "getApproveInstructionDataEncoder", {
  enumerable: true,
  get: function () { return token2022.getApproveInstructionDataEncoder; }
});
Object.defineProperty(exports, "getAssociatedTokenErrorMessage", {
  enumerable: true,
  get: function () { return token2022.getAssociatedTokenErrorMessage; }
});
Object.defineProperty(exports, "getAuthorityTypeCodec", {
  enumerable: true,
  get: function () { return token2022.getAuthorityTypeCodec; }
});
Object.defineProperty(exports, "getAuthorityTypeDecoder", {
  enumerable: true,
  get: function () { return token2022.getAuthorityTypeDecoder; }
});
Object.defineProperty(exports, "getAuthorityTypeEncoder", {
  enumerable: true,
  get: function () { return token2022.getAuthorityTypeEncoder; }
});
Object.defineProperty(exports, "getBurnCheckedDiscriminatorBytes", {
  enumerable: true,
  get: function () { return token2022.getBurnCheckedDiscriminatorBytes; }
});
Object.defineProperty(exports, "getBurnCheckedInstruction", {
  enumerable: true,
  get: function () { return token2022.getBurnCheckedInstruction; }
});
Object.defineProperty(exports, "getBurnCheckedInstructionDataCodec", {
  enumerable: true,
  get: function () { return token2022.getBurnCheckedInstructionDataCodec; }
});
Object.defineProperty(exports, "getBurnCheckedInstructionDataDecoder", {
  enumerable: true,
  get: function () { return token2022.getBurnCheckedInstructionDataDecoder; }
});
Object.defineProperty(exports, "getBurnCheckedInstructionDataEncoder", {
  enumerable: true,
  get: function () { return token2022.getBurnCheckedInstructionDataEncoder; }
});
Object.defineProperty(exports, "getBurnDiscriminatorBytes", {
  enumerable: true,
  get: function () { return token2022.getBurnDiscriminatorBytes; }
});
Object.defineProperty(exports, "getBurnInstruction", {
  enumerable: true,
  get: function () { return token2022.getBurnInstruction; }
});
Object.defineProperty(exports, "getBurnInstructionDataCodec", {
  enumerable: true,
  get: function () { return token2022.getBurnInstructionDataCodec; }
});
Object.defineProperty(exports, "getBurnInstructionDataDecoder", {
  enumerable: true,
  get: function () { return token2022.getBurnInstructionDataDecoder; }
});
Object.defineProperty(exports, "getBurnInstructionDataEncoder", {
  enumerable: true,
  get: function () { return token2022.getBurnInstructionDataEncoder; }
});
Object.defineProperty(exports, "getCloseAccountDiscriminatorBytes", {
  enumerable: true,
  get: function () { return token2022.getCloseAccountDiscriminatorBytes; }
});
Object.defineProperty(exports, "getCloseAccountInstruction", {
  enumerable: true,
  get: function () { return token2022.getCloseAccountInstruction; }
});
Object.defineProperty(exports, "getCloseAccountInstructionDataCodec", {
  enumerable: true,
  get: function () { return token2022.getCloseAccountInstructionDataCodec; }
});
Object.defineProperty(exports, "getCloseAccountInstructionDataDecoder", {
  enumerable: true,
  get: function () { return token2022.getCloseAccountInstructionDataDecoder; }
});
Object.defineProperty(exports, "getCloseAccountInstructionDataEncoder", {
  enumerable: true,
  get: function () { return token2022.getCloseAccountInstructionDataEncoder; }
});
Object.defineProperty(exports, "getConfidentialDepositConfidentialTransferDiscriminatorBytes", {
  enumerable: true,
  get: function () { return token2022.getConfidentialDepositConfidentialTransferDiscriminatorBytes; }
});
Object.defineProperty(exports, "getConfidentialDepositDiscriminatorBytes", {
  enumerable: true,
  get: function () { return token2022.getConfidentialDepositDiscriminatorBytes; }
});
Object.defineProperty(exports, "getConfidentialDepositInstruction", {
  enumerable: true,
  get: function () { return token2022.getConfidentialDepositInstruction; }
});
Object.defineProperty(exports, "getConfidentialDepositInstructionDataCodec", {
  enumerable: true,
  get: function () { return token2022.getConfidentialDepositInstructionDataCodec; }
});
Object.defineProperty(exports, "getConfidentialDepositInstructionDataDecoder", {
  enumerable: true,
  get: function () { return token2022.getConfidentialDepositInstructionDataDecoder; }
});
Object.defineProperty(exports, "getConfidentialDepositInstructionDataEncoder", {
  enumerable: true,
  get: function () { return token2022.getConfidentialDepositInstructionDataEncoder; }
});
Object.defineProperty(exports, "getConfidentialTransferConfidentialTransferDiscriminatorBytes", {
  enumerable: true,
  get: function () { return token2022.getConfidentialTransferConfidentialTransferDiscriminatorBytes; }
});
Object.defineProperty(exports, "getConfidentialTransferDiscriminatorBytes", {
  enumerable: true,
  get: function () { return token2022.getConfidentialTransferDiscriminatorBytes; }
});
Object.defineProperty(exports, "getConfidentialTransferInstruction", {
  enumerable: true,
  get: function () { return token2022.getConfidentialTransferInstruction; }
});
Object.defineProperty(exports, "getConfidentialTransferInstructionDataCodec", {
  enumerable: true,
  get: function () { return token2022.getConfidentialTransferInstructionDataCodec; }
});
Object.defineProperty(exports, "getConfidentialTransferInstructionDataDecoder", {
  enumerable: true,
  get: function () { return token2022.getConfidentialTransferInstructionDataDecoder; }
});
Object.defineProperty(exports, "getConfidentialTransferInstructionDataEncoder", {
  enumerable: true,
  get: function () { return token2022.getConfidentialTransferInstructionDataEncoder; }
});
Object.defineProperty(exports, "getConfidentialTransferWithFeeConfidentialTransferDiscriminatorBytes", {
  enumerable: true,
  get: function () { return token2022.getConfidentialTransferWithFeeConfidentialTransferDiscriminatorBytes; }
});
Object.defineProperty(exports, "getConfidentialTransferWithFeeDiscriminatorBytes", {
  enumerable: true,
  get: function () { return token2022.getConfidentialTransferWithFeeDiscriminatorBytes; }
});
Object.defineProperty(exports, "getConfidentialTransferWithFeeInstruction", {
  enumerable: true,
  get: function () { return token2022.getConfidentialTransferWithFeeInstruction; }
});
Object.defineProperty(exports, "getConfidentialTransferWithFeeInstructionDataCodec", {
  enumerable: true,
  get: function () { return token2022.getConfidentialTransferWithFeeInstructionDataCodec; }
});
Object.defineProperty(exports, "getConfidentialTransferWithFeeInstructionDataDecoder", {
  enumerable: true,
  get: function () { return token2022.getConfidentialTransferWithFeeInstructionDataDecoder; }
});
Object.defineProperty(exports, "getConfidentialTransferWithFeeInstructionDataEncoder", {
  enumerable: true,
  get: function () { return token2022.getConfidentialTransferWithFeeInstructionDataEncoder; }
});
Object.defineProperty(exports, "getConfidentialWithdrawConfidentialTransferDiscriminatorBytes", {
  enumerable: true,
  get: function () { return token2022.getConfidentialWithdrawConfidentialTransferDiscriminatorBytes; }
});
Object.defineProperty(exports, "getConfidentialWithdrawDiscriminatorBytes", {
  enumerable: true,
  get: function () { return token2022.getConfidentialWithdrawDiscriminatorBytes; }
});
Object.defineProperty(exports, "getConfidentialWithdrawInstruction", {
  enumerable: true,
  get: function () { return token2022.getConfidentialWithdrawInstruction; }
});
Object.defineProperty(exports, "getConfidentialWithdrawInstructionDataCodec", {
  enumerable: true,
  get: function () { return token2022.getConfidentialWithdrawInstructionDataCodec; }
});
Object.defineProperty(exports, "getConfidentialWithdrawInstructionDataDecoder", {
  enumerable: true,
  get: function () { return token2022.getConfidentialWithdrawInstructionDataDecoder; }
});
Object.defineProperty(exports, "getConfidentialWithdrawInstructionDataEncoder", {
  enumerable: true,
  get: function () { return token2022.getConfidentialWithdrawInstructionDataEncoder; }
});
Object.defineProperty(exports, "getConfigureConfidentialTransferAccountConfidentialTransferDiscriminatorBytes", {
  enumerable: true,
  get: function () { return token2022.getConfigureConfidentialTransferAccountConfidentialTransferDiscriminatorBytes; }
});
Object.defineProperty(exports, "getConfigureConfidentialTransferAccountDiscriminatorBytes", {
  enumerable: true,
  get: function () { return token2022.getConfigureConfidentialTransferAccountDiscriminatorBytes; }
});
Object.defineProperty(exports, "getConfigureConfidentialTransferAccountInstruction", {
  enumerable: true,
  get: function () { return token2022.getConfigureConfidentialTransferAccountInstruction; }
});
Object.defineProperty(exports, "getConfigureConfidentialTransferAccountInstructionDataCodec", {
  enumerable: true,
  get: function () { return token2022.getConfigureConfidentialTransferAccountInstructionDataCodec; }
});
Object.defineProperty(exports, "getConfigureConfidentialTransferAccountInstructionDataDecoder", {
  enumerable: true,
  get: function () { return token2022.getConfigureConfidentialTransferAccountInstructionDataDecoder; }
});
Object.defineProperty(exports, "getConfigureConfidentialTransferAccountInstructionDataEncoder", {
  enumerable: true,
  get: function () { return token2022.getConfigureConfidentialTransferAccountInstructionDataEncoder; }
});
Object.defineProperty(exports, "getCreateAssociatedTokenDiscriminatorBytes", {
  enumerable: true,
  get: function () { return token2022.getCreateAssociatedTokenDiscriminatorBytes; }
});
Object.defineProperty(exports, "getCreateAssociatedTokenIdempotentDiscriminatorBytes", {
  enumerable: true,
  get: function () { return token2022.getCreateAssociatedTokenIdempotentDiscriminatorBytes; }
});
Object.defineProperty(exports, "getCreateAssociatedTokenIdempotentInstruction", {
  enumerable: true,
  get: function () { return token2022.getCreateAssociatedTokenIdempotentInstruction; }
});
Object.defineProperty(exports, "getCreateAssociatedTokenIdempotentInstructionAsync", {
  enumerable: true,
  get: function () { return token2022.getCreateAssociatedTokenIdempotentInstructionAsync; }
});
Object.defineProperty(exports, "getCreateAssociatedTokenIdempotentInstructionDataCodec", {
  enumerable: true,
  get: function () { return token2022.getCreateAssociatedTokenIdempotentInstructionDataCodec; }
});
Object.defineProperty(exports, "getCreateAssociatedTokenIdempotentInstructionDataDecoder", {
  enumerable: true,
  get: function () { return token2022.getCreateAssociatedTokenIdempotentInstructionDataDecoder; }
});
Object.defineProperty(exports, "getCreateAssociatedTokenIdempotentInstructionDataEncoder", {
  enumerable: true,
  get: function () { return token2022.getCreateAssociatedTokenIdempotentInstructionDataEncoder; }
});
Object.defineProperty(exports, "getCreateAssociatedTokenInstruction", {
  enumerable: true,
  get: function () { return token2022.getCreateAssociatedTokenInstruction; }
});
Object.defineProperty(exports, "getCreateAssociatedTokenInstructionAsync", {
  enumerable: true,
  get: function () { return token2022.getCreateAssociatedTokenInstructionAsync; }
});
Object.defineProperty(exports, "getCreateAssociatedTokenInstructionDataCodec", {
  enumerable: true,
  get: function () { return token2022.getCreateAssociatedTokenInstructionDataCodec; }
});
Object.defineProperty(exports, "getCreateAssociatedTokenInstructionDataDecoder", {
  enumerable: true,
  get: function () { return token2022.getCreateAssociatedTokenInstructionDataDecoder; }
});
Object.defineProperty(exports, "getCreateAssociatedTokenInstructionDataEncoder", {
  enumerable: true,
  get: function () { return token2022.getCreateAssociatedTokenInstructionDataEncoder; }
});
Object.defineProperty(exports, "getCreateNativeMintDiscriminatorBytes", {
  enumerable: true,
  get: function () { return token2022.getCreateNativeMintDiscriminatorBytes; }
});
Object.defineProperty(exports, "getCreateNativeMintInstruction", {
  enumerable: true,
  get: function () { return token2022.getCreateNativeMintInstruction; }
});
Object.defineProperty(exports, "getCreateNativeMintInstructionDataCodec", {
  enumerable: true,
  get: function () { return token2022.getCreateNativeMintInstructionDataCodec; }
});
Object.defineProperty(exports, "getCreateNativeMintInstructionDataDecoder", {
  enumerable: true,
  get: function () { return token2022.getCreateNativeMintInstructionDataDecoder; }
});
Object.defineProperty(exports, "getCreateNativeMintInstructionDataEncoder", {
  enumerable: true,
  get: function () { return token2022.getCreateNativeMintInstructionDataEncoder; }
});
Object.defineProperty(exports, "getDecryptableBalanceCodec", {
  enumerable: true,
  get: function () { return token2022.getDecryptableBalanceCodec; }
});
Object.defineProperty(exports, "getDecryptableBalanceDecoder", {
  enumerable: true,
  get: function () { return token2022.getDecryptableBalanceDecoder; }
});
Object.defineProperty(exports, "getDecryptableBalanceEncoder", {
  enumerable: true,
  get: function () { return token2022.getDecryptableBalanceEncoder; }
});
Object.defineProperty(exports, "getDisableConfidentialCreditsConfidentialTransferDiscriminatorBytes", {
  enumerable: true,
  get: function () { return token2022.getDisableConfidentialCreditsConfidentialTransferDiscriminatorBytes; }
});
Object.defineProperty(exports, "getDisableConfidentialCreditsDiscriminatorBytes", {
  enumerable: true,
  get: function () { return token2022.getDisableConfidentialCreditsDiscriminatorBytes; }
});
Object.defineProperty(exports, "getDisableConfidentialCreditsInstruction", {
  enumerable: true,
  get: function () { return token2022.getDisableConfidentialCreditsInstruction; }
});
Object.defineProperty(exports, "getDisableConfidentialCreditsInstructionDataCodec", {
  enumerable: true,
  get: function () { return token2022.getDisableConfidentialCreditsInstructionDataCodec; }
});
Object.defineProperty(exports, "getDisableConfidentialCreditsInstructionDataDecoder", {
  enumerable: true,
  get: function () { return token2022.getDisableConfidentialCreditsInstructionDataDecoder; }
});
Object.defineProperty(exports, "getDisableConfidentialCreditsInstructionDataEncoder", {
  enumerable: true,
  get: function () { return token2022.getDisableConfidentialCreditsInstructionDataEncoder; }
});
Object.defineProperty(exports, "getDisableCpiGuardCpiGuardDiscriminatorBytes", {
  enumerable: true,
  get: function () { return token2022.getDisableCpiGuardCpiGuardDiscriminatorBytes; }
});
Object.defineProperty(exports, "getDisableCpiGuardDiscriminatorBytes", {
  enumerable: true,
  get: function () { return token2022.getDisableCpiGuardDiscriminatorBytes; }
});
Object.defineProperty(exports, "getDisableCpiGuardInstruction", {
  enumerable: true,
  get: function () { return token2022.getDisableCpiGuardInstruction; }
});
Object.defineProperty(exports, "getDisableCpiGuardInstructionDataCodec", {
  enumerable: true,
  get: function () { return token2022.getDisableCpiGuardInstructionDataCodec; }
});
Object.defineProperty(exports, "getDisableCpiGuardInstructionDataDecoder", {
  enumerable: true,
  get: function () { return token2022.getDisableCpiGuardInstructionDataDecoder; }
});
Object.defineProperty(exports, "getDisableCpiGuardInstructionDataEncoder", {
  enumerable: true,
  get: function () { return token2022.getDisableCpiGuardInstructionDataEncoder; }
});
Object.defineProperty(exports, "getDisableHarvestToMintConfidentialTransferFeeDiscriminatorBytes", {
  enumerable: true,
  get: function () { return token2022.getDisableHarvestToMintConfidentialTransferFeeDiscriminatorBytes; }
});
Object.defineProperty(exports, "getDisableHarvestToMintDiscriminatorBytes", {
  enumerable: true,
  get: function () { return token2022.getDisableHarvestToMintDiscriminatorBytes; }
});
Object.defineProperty(exports, "getDisableHarvestToMintInstruction", {
  enumerable: true,
  get: function () { return token2022.getDisableHarvestToMintInstruction; }
});
Object.defineProperty(exports, "getDisableHarvestToMintInstructionDataCodec", {
  enumerable: true,
  get: function () { return token2022.getDisableHarvestToMintInstructionDataCodec; }
});
Object.defineProperty(exports, "getDisableHarvestToMintInstructionDataDecoder", {
  enumerable: true,
  get: function () { return token2022.getDisableHarvestToMintInstructionDataDecoder; }
});
Object.defineProperty(exports, "getDisableHarvestToMintInstructionDataEncoder", {
  enumerable: true,
  get: function () { return token2022.getDisableHarvestToMintInstructionDataEncoder; }
});
Object.defineProperty(exports, "getDisableMemoTransfersDiscriminatorBytes", {
  enumerable: true,
  get: function () { return token2022.getDisableMemoTransfersDiscriminatorBytes; }
});
Object.defineProperty(exports, "getDisableMemoTransfersInstruction", {
  enumerable: true,
  get: function () { return token2022.getDisableMemoTransfersInstruction; }
});
Object.defineProperty(exports, "getDisableMemoTransfersInstructionDataCodec", {
  enumerable: true,
  get: function () { return token2022.getDisableMemoTransfersInstructionDataCodec; }
});
Object.defineProperty(exports, "getDisableMemoTransfersInstructionDataDecoder", {
  enumerable: true,
  get: function () { return token2022.getDisableMemoTransfersInstructionDataDecoder; }
});
Object.defineProperty(exports, "getDisableMemoTransfersInstructionDataEncoder", {
  enumerable: true,
  get: function () { return token2022.getDisableMemoTransfersInstructionDataEncoder; }
});
Object.defineProperty(exports, "getDisableMemoTransfersMemoTransfersDiscriminatorBytes", {
  enumerable: true,
  get: function () { return token2022.getDisableMemoTransfersMemoTransfersDiscriminatorBytes; }
});
Object.defineProperty(exports, "getDisableNonConfidentialCreditsConfidentialTransferDiscriminatorBytes", {
  enumerable: true,
  get: function () { return token2022.getDisableNonConfidentialCreditsConfidentialTransferDiscriminatorBytes; }
});
Object.defineProperty(exports, "getDisableNonConfidentialCreditsDiscriminatorBytes", {
  enumerable: true,
  get: function () { return token2022.getDisableNonConfidentialCreditsDiscriminatorBytes; }
});
Object.defineProperty(exports, "getDisableNonConfidentialCreditsInstruction", {
  enumerable: true,
  get: function () { return token2022.getDisableNonConfidentialCreditsInstruction; }
});
Object.defineProperty(exports, "getDisableNonConfidentialCreditsInstructionDataCodec", {
  enumerable: true,
  get: function () { return token2022.getDisableNonConfidentialCreditsInstructionDataCodec; }
});
Object.defineProperty(exports, "getDisableNonConfidentialCreditsInstructionDataDecoder", {
  enumerable: true,
  get: function () { return token2022.getDisableNonConfidentialCreditsInstructionDataDecoder; }
});
Object.defineProperty(exports, "getDisableNonConfidentialCreditsInstructionDataEncoder", {
  enumerable: true,
  get: function () { return token2022.getDisableNonConfidentialCreditsInstructionDataEncoder; }
});
Object.defineProperty(exports, "getEmitTokenMetadataDiscriminatorBytes", {
  enumerable: true,
  get: function () { return token2022.getEmitTokenMetadataDiscriminatorBytes; }
});
Object.defineProperty(exports, "getEmitTokenMetadataInstruction", {
  enumerable: true,
  get: function () { return token2022.getEmitTokenMetadataInstruction; }
});
Object.defineProperty(exports, "getEmitTokenMetadataInstructionDataCodec", {
  enumerable: true,
  get: function () { return token2022.getEmitTokenMetadataInstructionDataCodec; }
});
Object.defineProperty(exports, "getEmitTokenMetadataInstructionDataDecoder", {
  enumerable: true,
  get: function () { return token2022.getEmitTokenMetadataInstructionDataDecoder; }
});
Object.defineProperty(exports, "getEmitTokenMetadataInstructionDataEncoder", {
  enumerable: true,
  get: function () { return token2022.getEmitTokenMetadataInstructionDataEncoder; }
});
Object.defineProperty(exports, "getEmptyConfidentialTransferAccountConfidentialTransferDiscriminatorBytes", {
  enumerable: true,
  get: function () { return token2022.getEmptyConfidentialTransferAccountConfidentialTransferDiscriminatorBytes; }
});
Object.defineProperty(exports, "getEmptyConfidentialTransferAccountDiscriminatorBytes", {
  enumerable: true,
  get: function () { return token2022.getEmptyConfidentialTransferAccountDiscriminatorBytes; }
});
Object.defineProperty(exports, "getEmptyConfidentialTransferAccountInstruction", {
  enumerable: true,
  get: function () { return token2022.getEmptyConfidentialTransferAccountInstruction; }
});
Object.defineProperty(exports, "getEmptyConfidentialTransferAccountInstructionDataCodec", {
  enumerable: true,
  get: function () { return token2022.getEmptyConfidentialTransferAccountInstructionDataCodec; }
});
Object.defineProperty(exports, "getEmptyConfidentialTransferAccountInstructionDataDecoder", {
  enumerable: true,
  get: function () { return token2022.getEmptyConfidentialTransferAccountInstructionDataDecoder; }
});
Object.defineProperty(exports, "getEmptyConfidentialTransferAccountInstructionDataEncoder", {
  enumerable: true,
  get: function () { return token2022.getEmptyConfidentialTransferAccountInstructionDataEncoder; }
});
Object.defineProperty(exports, "getEnableConfidentialCreditsConfidentialTransferDiscriminatorBytes", {
  enumerable: true,
  get: function () { return token2022.getEnableConfidentialCreditsConfidentialTransferDiscriminatorBytes; }
});
Object.defineProperty(exports, "getEnableConfidentialCreditsDiscriminatorBytes", {
  enumerable: true,
  get: function () { return token2022.getEnableConfidentialCreditsDiscriminatorBytes; }
});
Object.defineProperty(exports, "getEnableConfidentialCreditsInstruction", {
  enumerable: true,
  get: function () { return token2022.getEnableConfidentialCreditsInstruction; }
});
Object.defineProperty(exports, "getEnableConfidentialCreditsInstructionDataCodec", {
  enumerable: true,
  get: function () { return token2022.getEnableConfidentialCreditsInstructionDataCodec; }
});
Object.defineProperty(exports, "getEnableConfidentialCreditsInstructionDataDecoder", {
  enumerable: true,
  get: function () { return token2022.getEnableConfidentialCreditsInstructionDataDecoder; }
});
Object.defineProperty(exports, "getEnableConfidentialCreditsInstructionDataEncoder", {
  enumerable: true,
  get: function () { return token2022.getEnableConfidentialCreditsInstructionDataEncoder; }
});
Object.defineProperty(exports, "getEnableCpiGuardCpiGuardDiscriminatorBytes", {
  enumerable: true,
  get: function () { return token2022.getEnableCpiGuardCpiGuardDiscriminatorBytes; }
});
Object.defineProperty(exports, "getEnableCpiGuardDiscriminatorBytes", {
  enumerable: true,
  get: function () { return token2022.getEnableCpiGuardDiscriminatorBytes; }
});
Object.defineProperty(exports, "getEnableCpiGuardInstruction", {
  enumerable: true,
  get: function () { return token2022.getEnableCpiGuardInstruction; }
});
Object.defineProperty(exports, "getEnableCpiGuardInstructionDataCodec", {
  enumerable: true,
  get: function () { return token2022.getEnableCpiGuardInstructionDataCodec; }
});
Object.defineProperty(exports, "getEnableCpiGuardInstructionDataDecoder", {
  enumerable: true,
  get: function () { return token2022.getEnableCpiGuardInstructionDataDecoder; }
});
Object.defineProperty(exports, "getEnableCpiGuardInstructionDataEncoder", {
  enumerable: true,
  get: function () { return token2022.getEnableCpiGuardInstructionDataEncoder; }
});
Object.defineProperty(exports, "getEnableHarvestToMintConfidentialTransferFeeDiscriminatorBytes", {
  enumerable: true,
  get: function () { return token2022.getEnableHarvestToMintConfidentialTransferFeeDiscriminatorBytes; }
});
Object.defineProperty(exports, "getEnableHarvestToMintDiscriminatorBytes", {
  enumerable: true,
  get: function () { return token2022.getEnableHarvestToMintDiscriminatorBytes; }
});
Object.defineProperty(exports, "getEnableHarvestToMintInstruction", {
  enumerable: true,
  get: function () { return token2022.getEnableHarvestToMintInstruction; }
});
Object.defineProperty(exports, "getEnableHarvestToMintInstructionDataCodec", {
  enumerable: true,
  get: function () { return token2022.getEnableHarvestToMintInstructionDataCodec; }
});
Object.defineProperty(exports, "getEnableHarvestToMintInstructionDataDecoder", {
  enumerable: true,
  get: function () { return token2022.getEnableHarvestToMintInstructionDataDecoder; }
});
Object.defineProperty(exports, "getEnableHarvestToMintInstructionDataEncoder", {
  enumerable: true,
  get: function () { return token2022.getEnableHarvestToMintInstructionDataEncoder; }
});
Object.defineProperty(exports, "getEnableMemoTransfersDiscriminatorBytes", {
  enumerable: true,
  get: function () { return token2022.getEnableMemoTransfersDiscriminatorBytes; }
});
Object.defineProperty(exports, "getEnableMemoTransfersInstruction", {
  enumerable: true,
  get: function () { return token2022.getEnableMemoTransfersInstruction; }
});
Object.defineProperty(exports, "getEnableMemoTransfersInstructionDataCodec", {
  enumerable: true,
  get: function () { return token2022.getEnableMemoTransfersInstructionDataCodec; }
});
Object.defineProperty(exports, "getEnableMemoTransfersInstructionDataDecoder", {
  enumerable: true,
  get: function () { return token2022.getEnableMemoTransfersInstructionDataDecoder; }
});
Object.defineProperty(exports, "getEnableMemoTransfersInstructionDataEncoder", {
  enumerable: true,
  get: function () { return token2022.getEnableMemoTransfersInstructionDataEncoder; }
});
Object.defineProperty(exports, "getEnableMemoTransfersMemoTransfersDiscriminatorBytes", {
  enumerable: true,
  get: function () { return token2022.getEnableMemoTransfersMemoTransfersDiscriminatorBytes; }
});
Object.defineProperty(exports, "getEnableNonConfidentialCreditsConfidentialTransferDiscriminatorBytes", {
  enumerable: true,
  get: function () { return token2022.getEnableNonConfidentialCreditsConfidentialTransferDiscriminatorBytes; }
});
Object.defineProperty(exports, "getEnableNonConfidentialCreditsDiscriminatorBytes", {
  enumerable: true,
  get: function () { return token2022.getEnableNonConfidentialCreditsDiscriminatorBytes; }
});
Object.defineProperty(exports, "getEnableNonConfidentialCreditsInstruction", {
  enumerable: true,
  get: function () { return token2022.getEnableNonConfidentialCreditsInstruction; }
});
Object.defineProperty(exports, "getEnableNonConfidentialCreditsInstructionDataCodec", {
  enumerable: true,
  get: function () { return token2022.getEnableNonConfidentialCreditsInstructionDataCodec; }
});
Object.defineProperty(exports, "getEnableNonConfidentialCreditsInstructionDataDecoder", {
  enumerable: true,
  get: function () { return token2022.getEnableNonConfidentialCreditsInstructionDataDecoder; }
});
Object.defineProperty(exports, "getEnableNonConfidentialCreditsInstructionDataEncoder", {
  enumerable: true,
  get: function () { return token2022.getEnableNonConfidentialCreditsInstructionDataEncoder; }
});
Object.defineProperty(exports, "getEncryptedBalanceCodec", {
  enumerable: true,
  get: function () { return token2022.getEncryptedBalanceCodec; }
});
Object.defineProperty(exports, "getEncryptedBalanceDecoder", {
  enumerable: true,
  get: function () { return token2022.getEncryptedBalanceDecoder; }
});
Object.defineProperty(exports, "getEncryptedBalanceEncoder", {
  enumerable: true,
  get: function () { return token2022.getEncryptedBalanceEncoder; }
});
Object.defineProperty(exports, "getExtensionCodec", {
  enumerable: true,
  get: function () { return token2022.getExtensionCodec; }
});
Object.defineProperty(exports, "getExtensionDecoder", {
  enumerable: true,
  get: function () { return token2022.getExtensionDecoder; }
});
Object.defineProperty(exports, "getExtensionEncoder", {
  enumerable: true,
  get: function () { return token2022.getExtensionEncoder; }
});
Object.defineProperty(exports, "getExtensionTypeCodec", {
  enumerable: true,
  get: function () { return token2022.getExtensionTypeCodec; }
});
Object.defineProperty(exports, "getExtensionTypeDecoder", {
  enumerable: true,
  get: function () { return token2022.getExtensionTypeDecoder; }
});
Object.defineProperty(exports, "getExtensionTypeEncoder", {
  enumerable: true,
  get: function () { return token2022.getExtensionTypeEncoder; }
});
Object.defineProperty(exports, "getFreezeAccountDiscriminatorBytes", {
  enumerable: true,
  get: function () { return token2022.getFreezeAccountDiscriminatorBytes; }
});
Object.defineProperty(exports, "getFreezeAccountInstruction", {
  enumerable: true,
  get: function () { return token2022.getFreezeAccountInstruction; }
});
Object.defineProperty(exports, "getFreezeAccountInstructionDataCodec", {
  enumerable: true,
  get: function () { return token2022.getFreezeAccountInstructionDataCodec; }
});
Object.defineProperty(exports, "getFreezeAccountInstructionDataDecoder", {
  enumerable: true,
  get: function () { return token2022.getFreezeAccountInstructionDataDecoder; }
});
Object.defineProperty(exports, "getFreezeAccountInstructionDataEncoder", {
  enumerable: true,
  get: function () { return token2022.getFreezeAccountInstructionDataEncoder; }
});
Object.defineProperty(exports, "getGetAccountDataSizeDiscriminatorBytes", {
  enumerable: true,
  get: function () { return token2022.getGetAccountDataSizeDiscriminatorBytes; }
});
Object.defineProperty(exports, "getGetAccountDataSizeInstruction", {
  enumerable: true,
  get: function () { return token2022.getGetAccountDataSizeInstruction; }
});
Object.defineProperty(exports, "getGetAccountDataSizeInstructionDataCodec", {
  enumerable: true,
  get: function () { return token2022.getGetAccountDataSizeInstructionDataCodec; }
});
Object.defineProperty(exports, "getGetAccountDataSizeInstructionDataDecoder", {
  enumerable: true,
  get: function () { return token2022.getGetAccountDataSizeInstructionDataDecoder; }
});
Object.defineProperty(exports, "getGetAccountDataSizeInstructionDataEncoder", {
  enumerable: true,
  get: function () { return token2022.getGetAccountDataSizeInstructionDataEncoder; }
});
Object.defineProperty(exports, "getHarvestWithheldTokensToMintDiscriminatorBytes", {
  enumerable: true,
  get: function () { return token2022.getHarvestWithheldTokensToMintDiscriminatorBytes; }
});
Object.defineProperty(exports, "getHarvestWithheldTokensToMintForConfidentialTransferFeeConfidentialTransferFeeDiscriminatorBytes", {
  enumerable: true,
  get: function () { return token2022.getHarvestWithheldTokensToMintForConfidentialTransferFeeConfidentialTransferFeeDiscriminatorBytes; }
});
Object.defineProperty(exports, "getHarvestWithheldTokensToMintForConfidentialTransferFeeDiscriminatorBytes", {
  enumerable: true,
  get: function () { return token2022.getHarvestWithheldTokensToMintForConfidentialTransferFeeDiscriminatorBytes; }
});
Object.defineProperty(exports, "getHarvestWithheldTokensToMintForConfidentialTransferFeeInstruction", {
  enumerable: true,
  get: function () { return token2022.getHarvestWithheldTokensToMintForConfidentialTransferFeeInstruction; }
});
Object.defineProperty(exports, "getHarvestWithheldTokensToMintForConfidentialTransferFeeInstructionDataCodec", {
  enumerable: true,
  get: function () { return token2022.getHarvestWithheldTokensToMintForConfidentialTransferFeeInstructionDataCodec; }
});
Object.defineProperty(exports, "getHarvestWithheldTokensToMintForConfidentialTransferFeeInstructionDataDecoder", {
  enumerable: true,
  get: function () { return token2022.getHarvestWithheldTokensToMintForConfidentialTransferFeeInstructionDataDecoder; }
});
Object.defineProperty(exports, "getHarvestWithheldTokensToMintForConfidentialTransferFeeInstructionDataEncoder", {
  enumerable: true,
  get: function () { return token2022.getHarvestWithheldTokensToMintForConfidentialTransferFeeInstructionDataEncoder; }
});
Object.defineProperty(exports, "getHarvestWithheldTokensToMintInstruction", {
  enumerable: true,
  get: function () { return token2022.getHarvestWithheldTokensToMintInstruction; }
});
Object.defineProperty(exports, "getHarvestWithheldTokensToMintInstructionDataCodec", {
  enumerable: true,
  get: function () { return token2022.getHarvestWithheldTokensToMintInstructionDataCodec; }
});
Object.defineProperty(exports, "getHarvestWithheldTokensToMintInstructionDataDecoder", {
  enumerable: true,
  get: function () { return token2022.getHarvestWithheldTokensToMintInstructionDataDecoder; }
});
Object.defineProperty(exports, "getHarvestWithheldTokensToMintInstructionDataEncoder", {
  enumerable: true,
  get: function () { return token2022.getHarvestWithheldTokensToMintInstructionDataEncoder; }
});
Object.defineProperty(exports, "getHarvestWithheldTokensToMintTransferFeeDiscriminatorBytes", {
  enumerable: true,
  get: function () { return token2022.getHarvestWithheldTokensToMintTransferFeeDiscriminatorBytes; }
});
Object.defineProperty(exports, "getInitializeAccount2DiscriminatorBytes", {
  enumerable: true,
  get: function () { return token2022.getInitializeAccount2DiscriminatorBytes; }
});
Object.defineProperty(exports, "getInitializeAccount2Instruction", {
  enumerable: true,
  get: function () { return token2022.getInitializeAccount2Instruction; }
});
Object.defineProperty(exports, "getInitializeAccount2InstructionDataCodec", {
  enumerable: true,
  get: function () { return token2022.getInitializeAccount2InstructionDataCodec; }
});
Object.defineProperty(exports, "getInitializeAccount2InstructionDataDecoder", {
  enumerable: true,
  get: function () { return token2022.getInitializeAccount2InstructionDataDecoder; }
});
Object.defineProperty(exports, "getInitializeAccount2InstructionDataEncoder", {
  enumerable: true,
  get: function () { return token2022.getInitializeAccount2InstructionDataEncoder; }
});
Object.defineProperty(exports, "getInitializeAccount3DiscriminatorBytes", {
  enumerable: true,
  get: function () { return token2022.getInitializeAccount3DiscriminatorBytes; }
});
Object.defineProperty(exports, "getInitializeAccount3Instruction", {
  enumerable: true,
  get: function () { return token2022.getInitializeAccount3Instruction; }
});
Object.defineProperty(exports, "getInitializeAccount3InstructionDataCodec", {
  enumerable: true,
  get: function () { return token2022.getInitializeAccount3InstructionDataCodec; }
});
Object.defineProperty(exports, "getInitializeAccount3InstructionDataDecoder", {
  enumerable: true,
  get: function () { return token2022.getInitializeAccount3InstructionDataDecoder; }
});
Object.defineProperty(exports, "getInitializeAccount3InstructionDataEncoder", {
  enumerable: true,
  get: function () { return token2022.getInitializeAccount3InstructionDataEncoder; }
});
Object.defineProperty(exports, "getInitializeAccountDiscriminatorBytes", {
  enumerable: true,
  get: function () { return token2022.getInitializeAccountDiscriminatorBytes; }
});
Object.defineProperty(exports, "getInitializeAccountInstruction", {
  enumerable: true,
  get: function () { return token2022.getInitializeAccountInstruction; }
});
Object.defineProperty(exports, "getInitializeAccountInstructionDataCodec", {
  enumerable: true,
  get: function () { return token2022.getInitializeAccountInstructionDataCodec; }
});
Object.defineProperty(exports, "getInitializeAccountInstructionDataDecoder", {
  enumerable: true,
  get: function () { return token2022.getInitializeAccountInstructionDataDecoder; }
});
Object.defineProperty(exports, "getInitializeAccountInstructionDataEncoder", {
  enumerable: true,
  get: function () { return token2022.getInitializeAccountInstructionDataEncoder; }
});
Object.defineProperty(exports, "getInitializeConfidentialTransferFeeConfidentialTransferFeeDiscriminatorBytes", {
  enumerable: true,
  get: function () { return token2022.getInitializeConfidentialTransferFeeConfidentialTransferFeeDiscriminatorBytes; }
});
Object.defineProperty(exports, "getInitializeConfidentialTransferFeeDiscriminatorBytes", {
  enumerable: true,
  get: function () { return token2022.getInitializeConfidentialTransferFeeDiscriminatorBytes; }
});
Object.defineProperty(exports, "getInitializeConfidentialTransferFeeInstruction", {
  enumerable: true,
  get: function () { return token2022.getInitializeConfidentialTransferFeeInstruction; }
});
Object.defineProperty(exports, "getInitializeConfidentialTransferFeeInstructionDataCodec", {
  enumerable: true,
  get: function () { return token2022.getInitializeConfidentialTransferFeeInstructionDataCodec; }
});
Object.defineProperty(exports, "getInitializeConfidentialTransferFeeInstructionDataDecoder", {
  enumerable: true,
  get: function () { return token2022.getInitializeConfidentialTransferFeeInstructionDataDecoder; }
});
Object.defineProperty(exports, "getInitializeConfidentialTransferFeeInstructionDataEncoder", {
  enumerable: true,
  get: function () { return token2022.getInitializeConfidentialTransferFeeInstructionDataEncoder; }
});
Object.defineProperty(exports, "getInitializeConfidentialTransferMintConfidentialTransferDiscriminatorBytes", {
  enumerable: true,
  get: function () { return token2022.getInitializeConfidentialTransferMintConfidentialTransferDiscriminatorBytes; }
});
Object.defineProperty(exports, "getInitializeConfidentialTransferMintDiscriminatorBytes", {
  enumerable: true,
  get: function () { return token2022.getInitializeConfidentialTransferMintDiscriminatorBytes; }
});
Object.defineProperty(exports, "getInitializeConfidentialTransferMintInstruction", {
  enumerable: true,
  get: function () { return token2022.getInitializeConfidentialTransferMintInstruction; }
});
Object.defineProperty(exports, "getInitializeConfidentialTransferMintInstructionDataCodec", {
  enumerable: true,
  get: function () { return token2022.getInitializeConfidentialTransferMintInstructionDataCodec; }
});
Object.defineProperty(exports, "getInitializeConfidentialTransferMintInstructionDataDecoder", {
  enumerable: true,
  get: function () { return token2022.getInitializeConfidentialTransferMintInstructionDataDecoder; }
});
Object.defineProperty(exports, "getInitializeConfidentialTransferMintInstructionDataEncoder", {
  enumerable: true,
  get: function () { return token2022.getInitializeConfidentialTransferMintInstructionDataEncoder; }
});
Object.defineProperty(exports, "getInitializeDefaultAccountStateDefaultAccountStateDiscriminatorBytes", {
  enumerable: true,
  get: function () { return token2022.getInitializeDefaultAccountStateDefaultAccountStateDiscriminatorBytes; }
});
Object.defineProperty(exports, "getInitializeDefaultAccountStateDiscriminatorBytes", {
  enumerable: true,
  get: function () { return token2022.getInitializeDefaultAccountStateDiscriminatorBytes; }
});
Object.defineProperty(exports, "getInitializeDefaultAccountStateInstruction", {
  enumerable: true,
  get: function () { return token2022.getInitializeDefaultAccountStateInstruction; }
});
Object.defineProperty(exports, "getInitializeDefaultAccountStateInstructionDataCodec", {
  enumerable: true,
  get: function () { return token2022.getInitializeDefaultAccountStateInstructionDataCodec; }
});
Object.defineProperty(exports, "getInitializeDefaultAccountStateInstructionDataDecoder", {
  enumerable: true,
  get: function () { return token2022.getInitializeDefaultAccountStateInstructionDataDecoder; }
});
Object.defineProperty(exports, "getInitializeDefaultAccountStateInstructionDataEncoder", {
  enumerable: true,
  get: function () { return token2022.getInitializeDefaultAccountStateInstructionDataEncoder; }
});
Object.defineProperty(exports, "getInitializeGroupMemberPointerDiscriminatorBytes", {
  enumerable: true,
  get: function () { return token2022.getInitializeGroupMemberPointerDiscriminatorBytes; }
});
Object.defineProperty(exports, "getInitializeGroupMemberPointerGroupMemberPointerDiscriminatorBytes", {
  enumerable: true,
  get: function () { return token2022.getInitializeGroupMemberPointerGroupMemberPointerDiscriminatorBytes; }
});
Object.defineProperty(exports, "getInitializeGroupMemberPointerInstruction", {
  enumerable: true,
  get: function () { return token2022.getInitializeGroupMemberPointerInstruction; }
});
Object.defineProperty(exports, "getInitializeGroupMemberPointerInstructionDataCodec", {
  enumerable: true,
  get: function () { return token2022.getInitializeGroupMemberPointerInstructionDataCodec; }
});
Object.defineProperty(exports, "getInitializeGroupMemberPointerInstructionDataDecoder", {
  enumerable: true,
  get: function () { return token2022.getInitializeGroupMemberPointerInstructionDataDecoder; }
});
Object.defineProperty(exports, "getInitializeGroupMemberPointerInstructionDataEncoder", {
  enumerable: true,
  get: function () { return token2022.getInitializeGroupMemberPointerInstructionDataEncoder; }
});
Object.defineProperty(exports, "getInitializeGroupPointerDiscriminatorBytes", {
  enumerable: true,
  get: function () { return token2022.getInitializeGroupPointerDiscriminatorBytes; }
});
Object.defineProperty(exports, "getInitializeGroupPointerGroupPointerDiscriminatorBytes", {
  enumerable: true,
  get: function () { return token2022.getInitializeGroupPointerGroupPointerDiscriminatorBytes; }
});
Object.defineProperty(exports, "getInitializeGroupPointerInstruction", {
  enumerable: true,
  get: function () { return token2022.getInitializeGroupPointerInstruction; }
});
Object.defineProperty(exports, "getInitializeGroupPointerInstructionDataCodec", {
  enumerable: true,
  get: function () { return token2022.getInitializeGroupPointerInstructionDataCodec; }
});
Object.defineProperty(exports, "getInitializeGroupPointerInstructionDataDecoder", {
  enumerable: true,
  get: function () { return token2022.getInitializeGroupPointerInstructionDataDecoder; }
});
Object.defineProperty(exports, "getInitializeGroupPointerInstructionDataEncoder", {
  enumerable: true,
  get: function () { return token2022.getInitializeGroupPointerInstructionDataEncoder; }
});
Object.defineProperty(exports, "getInitializeImmutableOwnerDiscriminatorBytes", {
  enumerable: true,
  get: function () { return token2022.getInitializeImmutableOwnerDiscriminatorBytes; }
});
Object.defineProperty(exports, "getInitializeImmutableOwnerInstruction", {
  enumerable: true,
  get: function () { return token2022.getInitializeImmutableOwnerInstruction; }
});
Object.defineProperty(exports, "getInitializeImmutableOwnerInstructionDataCodec", {
  enumerable: true,
  get: function () { return token2022.getInitializeImmutableOwnerInstructionDataCodec; }
});
Object.defineProperty(exports, "getInitializeImmutableOwnerInstructionDataDecoder", {
  enumerable: true,
  get: function () { return token2022.getInitializeImmutableOwnerInstructionDataDecoder; }
});
Object.defineProperty(exports, "getInitializeImmutableOwnerInstructionDataEncoder", {
  enumerable: true,
  get: function () { return token2022.getInitializeImmutableOwnerInstructionDataEncoder; }
});
Object.defineProperty(exports, "getInitializeInterestBearingMintDiscriminatorBytes", {
  enumerable: true,
  get: function () { return token2022.getInitializeInterestBearingMintDiscriminatorBytes; }
});
Object.defineProperty(exports, "getInitializeInterestBearingMintInstruction", {
  enumerable: true,
  get: function () { return token2022.getInitializeInterestBearingMintInstruction; }
});
Object.defineProperty(exports, "getInitializeInterestBearingMintInstructionDataCodec", {
  enumerable: true,
  get: function () { return token2022.getInitializeInterestBearingMintInstructionDataCodec; }
});
Object.defineProperty(exports, "getInitializeInterestBearingMintInstructionDataDecoder", {
  enumerable: true,
  get: function () { return token2022.getInitializeInterestBearingMintInstructionDataDecoder; }
});
Object.defineProperty(exports, "getInitializeInterestBearingMintInstructionDataEncoder", {
  enumerable: true,
  get: function () { return token2022.getInitializeInterestBearingMintInstructionDataEncoder; }
});
Object.defineProperty(exports, "getInitializeInterestBearingMintInterestBearingMintDiscriminatorBytes", {
  enumerable: true,
  get: function () { return token2022.getInitializeInterestBearingMintInterestBearingMintDiscriminatorBytes; }
});
Object.defineProperty(exports, "getInitializeMetadataPointerDiscriminatorBytes", {
  enumerable: true,
  get: function () { return token2022.getInitializeMetadataPointerDiscriminatorBytes; }
});
Object.defineProperty(exports, "getInitializeMetadataPointerInstruction", {
  enumerable: true,
  get: function () { return token2022.getInitializeMetadataPointerInstruction; }
});
Object.defineProperty(exports, "getInitializeMetadataPointerInstructionDataCodec", {
  enumerable: true,
  get: function () { return token2022.getInitializeMetadataPointerInstructionDataCodec; }
});
Object.defineProperty(exports, "getInitializeMetadataPointerInstructionDataDecoder", {
  enumerable: true,
  get: function () { return token2022.getInitializeMetadataPointerInstructionDataDecoder; }
});
Object.defineProperty(exports, "getInitializeMetadataPointerInstructionDataEncoder", {
  enumerable: true,
  get: function () { return token2022.getInitializeMetadataPointerInstructionDataEncoder; }
});
Object.defineProperty(exports, "getInitializeMetadataPointerMetadataPointerDiscriminatorBytes", {
  enumerable: true,
  get: function () { return token2022.getInitializeMetadataPointerMetadataPointerDiscriminatorBytes; }
});
Object.defineProperty(exports, "getInitializeMint2DiscriminatorBytes", {
  enumerable: true,
  get: function () { return token2022.getInitializeMint2DiscriminatorBytes; }
});
Object.defineProperty(exports, "getInitializeMint2Instruction", {
  enumerable: true,
  get: function () { return token2022.getInitializeMint2Instruction; }
});
Object.defineProperty(exports, "getInitializeMint2InstructionDataCodec", {
  enumerable: true,
  get: function () { return token2022.getInitializeMint2InstructionDataCodec; }
});
Object.defineProperty(exports, "getInitializeMint2InstructionDataDecoder", {
  enumerable: true,
  get: function () { return token2022.getInitializeMint2InstructionDataDecoder; }
});
Object.defineProperty(exports, "getInitializeMint2InstructionDataEncoder", {
  enumerable: true,
  get: function () { return token2022.getInitializeMint2InstructionDataEncoder; }
});
Object.defineProperty(exports, "getInitializeMintCloseAuthorityDiscriminatorBytes", {
  enumerable: true,
  get: function () { return token2022.getInitializeMintCloseAuthorityDiscriminatorBytes; }
});
Object.defineProperty(exports, "getInitializeMintCloseAuthorityInstruction", {
  enumerable: true,
  get: function () { return token2022.getInitializeMintCloseAuthorityInstruction; }
});
Object.defineProperty(exports, "getInitializeMintCloseAuthorityInstructionDataCodec", {
  enumerable: true,
  get: function () { return token2022.getInitializeMintCloseAuthorityInstructionDataCodec; }
});
Object.defineProperty(exports, "getInitializeMintCloseAuthorityInstructionDataDecoder", {
  enumerable: true,
  get: function () { return token2022.getInitializeMintCloseAuthorityInstructionDataDecoder; }
});
Object.defineProperty(exports, "getInitializeMintCloseAuthorityInstructionDataEncoder", {
  enumerable: true,
  get: function () { return token2022.getInitializeMintCloseAuthorityInstructionDataEncoder; }
});
Object.defineProperty(exports, "getInitializeMintDiscriminatorBytes", {
  enumerable: true,
  get: function () { return token2022.getInitializeMintDiscriminatorBytes; }
});
Object.defineProperty(exports, "getInitializeMintInstruction", {
  enumerable: true,
  get: function () { return token2022.getInitializeMintInstruction; }
});
Object.defineProperty(exports, "getInitializeMintInstructionDataCodec", {
  enumerable: true,
  get: function () { return token2022.getInitializeMintInstructionDataCodec; }
});
Object.defineProperty(exports, "getInitializeMintInstructionDataDecoder", {
  enumerable: true,
  get: function () { return token2022.getInitializeMintInstructionDataDecoder; }
});
Object.defineProperty(exports, "getInitializeMintInstructionDataEncoder", {
  enumerable: true,
  get: function () { return token2022.getInitializeMintInstructionDataEncoder; }
});
Object.defineProperty(exports, "getInitializeMultisig2DiscriminatorBytes", {
  enumerable: true,
  get: function () { return token2022.getInitializeMultisig2DiscriminatorBytes; }
});
Object.defineProperty(exports, "getInitializeMultisig2Instruction", {
  enumerable: true,
  get: function () { return token2022.getInitializeMultisig2Instruction; }
});
Object.defineProperty(exports, "getInitializeMultisig2InstructionDataCodec", {
  enumerable: true,
  get: function () { return token2022.getInitializeMultisig2InstructionDataCodec; }
});
Object.defineProperty(exports, "getInitializeMultisig2InstructionDataDecoder", {
  enumerable: true,
  get: function () { return token2022.getInitializeMultisig2InstructionDataDecoder; }
});
Object.defineProperty(exports, "getInitializeMultisig2InstructionDataEncoder", {
  enumerable: true,
  get: function () { return token2022.getInitializeMultisig2InstructionDataEncoder; }
});
Object.defineProperty(exports, "getInitializeMultisigDiscriminatorBytes", {
  enumerable: true,
  get: function () { return token2022.getInitializeMultisigDiscriminatorBytes; }
});
Object.defineProperty(exports, "getInitializeMultisigInstruction", {
  enumerable: true,
  get: function () { return token2022.getInitializeMultisigInstruction; }
});
Object.defineProperty(exports, "getInitializeMultisigInstructionDataCodec", {
  enumerable: true,
  get: function () { return token2022.getInitializeMultisigInstructionDataCodec; }
});
Object.defineProperty(exports, "getInitializeMultisigInstructionDataDecoder", {
  enumerable: true,
  get: function () { return token2022.getInitializeMultisigInstructionDataDecoder; }
});
Object.defineProperty(exports, "getInitializeMultisigInstructionDataEncoder", {
  enumerable: true,
  get: function () { return token2022.getInitializeMultisigInstructionDataEncoder; }
});
Object.defineProperty(exports, "getInitializeNonTransferableMintDiscriminatorBytes", {
  enumerable: true,
  get: function () { return token2022.getInitializeNonTransferableMintDiscriminatorBytes; }
});
Object.defineProperty(exports, "getInitializeNonTransferableMintInstruction", {
  enumerable: true,
  get: function () { return token2022.getInitializeNonTransferableMintInstruction; }
});
Object.defineProperty(exports, "getInitializeNonTransferableMintInstructionDataCodec", {
  enumerable: true,
  get: function () { return token2022.getInitializeNonTransferableMintInstructionDataCodec; }
});
Object.defineProperty(exports, "getInitializeNonTransferableMintInstructionDataDecoder", {
  enumerable: true,
  get: function () { return token2022.getInitializeNonTransferableMintInstructionDataDecoder; }
});
Object.defineProperty(exports, "getInitializeNonTransferableMintInstructionDataEncoder", {
  enumerable: true,
  get: function () { return token2022.getInitializeNonTransferableMintInstructionDataEncoder; }
});
Object.defineProperty(exports, "getInitializePausableConfigDiscriminatorBytes", {
  enumerable: true,
  get: function () { return token2022.getInitializePausableConfigDiscriminatorBytes; }
});
Object.defineProperty(exports, "getInitializePausableConfigInstruction", {
  enumerable: true,
  get: function () { return token2022.getInitializePausableConfigInstruction; }
});
Object.defineProperty(exports, "getInitializePausableConfigInstructionDataCodec", {
  enumerable: true,
  get: function () { return token2022.getInitializePausableConfigInstructionDataCodec; }
});
Object.defineProperty(exports, "getInitializePausableConfigInstructionDataDecoder", {
  enumerable: true,
  get: function () { return token2022.getInitializePausableConfigInstructionDataDecoder; }
});
Object.defineProperty(exports, "getInitializePausableConfigInstructionDataEncoder", {
  enumerable: true,
  get: function () { return token2022.getInitializePausableConfigInstructionDataEncoder; }
});
Object.defineProperty(exports, "getInitializePausableConfigPausableDiscriminatorBytes", {
  enumerable: true,
  get: function () { return token2022.getInitializePausableConfigPausableDiscriminatorBytes; }
});
Object.defineProperty(exports, "getInitializePermanentDelegateDiscriminatorBytes", {
  enumerable: true,
  get: function () { return token2022.getInitializePermanentDelegateDiscriminatorBytes; }
});
Object.defineProperty(exports, "getInitializePermanentDelegateInstruction", {
  enumerable: true,
  get: function () { return token2022.getInitializePermanentDelegateInstruction; }
});
Object.defineProperty(exports, "getInitializePermanentDelegateInstructionDataCodec", {
  enumerable: true,
  get: function () { return token2022.getInitializePermanentDelegateInstructionDataCodec; }
});
Object.defineProperty(exports, "getInitializePermanentDelegateInstructionDataDecoder", {
  enumerable: true,
  get: function () { return token2022.getInitializePermanentDelegateInstructionDataDecoder; }
});
Object.defineProperty(exports, "getInitializePermanentDelegateInstructionDataEncoder", {
  enumerable: true,
  get: function () { return token2022.getInitializePermanentDelegateInstructionDataEncoder; }
});
Object.defineProperty(exports, "getInitializeScaledUiAmountMintDiscriminatorBytes", {
  enumerable: true,
  get: function () { return token2022.getInitializeScaledUiAmountMintDiscriminatorBytes; }
});
Object.defineProperty(exports, "getInitializeScaledUiAmountMintInstruction", {
  enumerable: true,
  get: function () { return token2022.getInitializeScaledUiAmountMintInstruction; }
});
Object.defineProperty(exports, "getInitializeScaledUiAmountMintInstructionDataCodec", {
  enumerable: true,
  get: function () { return token2022.getInitializeScaledUiAmountMintInstructionDataCodec; }
});
Object.defineProperty(exports, "getInitializeScaledUiAmountMintInstructionDataDecoder", {
  enumerable: true,
  get: function () { return token2022.getInitializeScaledUiAmountMintInstructionDataDecoder; }
});
Object.defineProperty(exports, "getInitializeScaledUiAmountMintInstructionDataEncoder", {
  enumerable: true,
  get: function () { return token2022.getInitializeScaledUiAmountMintInstructionDataEncoder; }
});
Object.defineProperty(exports, "getInitializeScaledUiAmountMintScaledUiAmountMintDiscriminatorBytes", {
  enumerable: true,
  get: function () { return token2022.getInitializeScaledUiAmountMintScaledUiAmountMintDiscriminatorBytes; }
});
Object.defineProperty(exports, "getInitializeTokenGroupDiscriminatorBytes", {
  enumerable: true,
  get: function () { return token2022.getInitializeTokenGroupDiscriminatorBytes; }
});
Object.defineProperty(exports, "getInitializeTokenGroupInstruction", {
  enumerable: true,
  get: function () { return token2022.getInitializeTokenGroupInstruction; }
});
Object.defineProperty(exports, "getInitializeTokenGroupInstructionDataCodec", {
  enumerable: true,
  get: function () { return token2022.getInitializeTokenGroupInstructionDataCodec; }
});
Object.defineProperty(exports, "getInitializeTokenGroupInstructionDataDecoder", {
  enumerable: true,
  get: function () { return token2022.getInitializeTokenGroupInstructionDataDecoder; }
});
Object.defineProperty(exports, "getInitializeTokenGroupInstructionDataEncoder", {
  enumerable: true,
  get: function () { return token2022.getInitializeTokenGroupInstructionDataEncoder; }
});
Object.defineProperty(exports, "getInitializeTokenGroupMemberDiscriminatorBytes", {
  enumerable: true,
  get: function () { return token2022.getInitializeTokenGroupMemberDiscriminatorBytes; }
});
Object.defineProperty(exports, "getInitializeTokenGroupMemberInstruction", {
  enumerable: true,
  get: function () { return token2022.getInitializeTokenGroupMemberInstruction; }
});
Object.defineProperty(exports, "getInitializeTokenGroupMemberInstructionDataCodec", {
  enumerable: true,
  get: function () { return token2022.getInitializeTokenGroupMemberInstructionDataCodec; }
});
Object.defineProperty(exports, "getInitializeTokenGroupMemberInstructionDataDecoder", {
  enumerable: true,
  get: function () { return token2022.getInitializeTokenGroupMemberInstructionDataDecoder; }
});
Object.defineProperty(exports, "getInitializeTokenGroupMemberInstructionDataEncoder", {
  enumerable: true,
  get: function () { return token2022.getInitializeTokenGroupMemberInstructionDataEncoder; }
});
Object.defineProperty(exports, "getInitializeTokenMetadataDiscriminatorBytes", {
  enumerable: true,
  get: function () { return token2022.getInitializeTokenMetadataDiscriminatorBytes; }
});
Object.defineProperty(exports, "getInitializeTokenMetadataInstruction", {
  enumerable: true,
  get: function () { return token2022.getInitializeTokenMetadataInstruction; }
});
Object.defineProperty(exports, "getInitializeTokenMetadataInstructionDataCodec", {
  enumerable: true,
  get: function () { return token2022.getInitializeTokenMetadataInstructionDataCodec; }
});
Object.defineProperty(exports, "getInitializeTokenMetadataInstructionDataDecoder", {
  enumerable: true,
  get: function () { return token2022.getInitializeTokenMetadataInstructionDataDecoder; }
});
Object.defineProperty(exports, "getInitializeTokenMetadataInstructionDataEncoder", {
  enumerable: true,
  get: function () { return token2022.getInitializeTokenMetadataInstructionDataEncoder; }
});
Object.defineProperty(exports, "getInitializeTransferFeeConfigDiscriminatorBytes", {
  enumerable: true,
  get: function () { return token2022.getInitializeTransferFeeConfigDiscriminatorBytes; }
});
Object.defineProperty(exports, "getInitializeTransferFeeConfigInstruction", {
  enumerable: true,
  get: function () { return token2022.getInitializeTransferFeeConfigInstruction; }
});
Object.defineProperty(exports, "getInitializeTransferFeeConfigInstructionDataCodec", {
  enumerable: true,
  get: function () { return token2022.getInitializeTransferFeeConfigInstructionDataCodec; }
});
Object.defineProperty(exports, "getInitializeTransferFeeConfigInstructionDataDecoder", {
  enumerable: true,
  get: function () { return token2022.getInitializeTransferFeeConfigInstructionDataDecoder; }
});
Object.defineProperty(exports, "getInitializeTransferFeeConfigInstructionDataEncoder", {
  enumerable: true,
  get: function () { return token2022.getInitializeTransferFeeConfigInstructionDataEncoder; }
});
Object.defineProperty(exports, "getInitializeTransferFeeConfigTransferFeeDiscriminatorBytes", {
  enumerable: true,
  get: function () { return token2022.getInitializeTransferFeeConfigTransferFeeDiscriminatorBytes; }
});
Object.defineProperty(exports, "getInitializeTransferHookDiscriminatorBytes", {
  enumerable: true,
  get: function () { return token2022.getInitializeTransferHookDiscriminatorBytes; }
});
Object.defineProperty(exports, "getInitializeTransferHookInstruction", {
  enumerable: true,
  get: function () { return token2022.getInitializeTransferHookInstruction; }
});
Object.defineProperty(exports, "getInitializeTransferHookInstructionDataCodec", {
  enumerable: true,
  get: function () { return token2022.getInitializeTransferHookInstructionDataCodec; }
});
Object.defineProperty(exports, "getInitializeTransferHookInstructionDataDecoder", {
  enumerable: true,
  get: function () { return token2022.getInitializeTransferHookInstructionDataDecoder; }
});
Object.defineProperty(exports, "getInitializeTransferHookInstructionDataEncoder", {
  enumerable: true,
  get: function () { return token2022.getInitializeTransferHookInstructionDataEncoder; }
});
Object.defineProperty(exports, "getInitializeTransferHookTransferHookDiscriminatorBytes", {
  enumerable: true,
  get: function () { return token2022.getInitializeTransferHookTransferHookDiscriminatorBytes; }
});
Object.defineProperty(exports, "getMintCodec", {
  enumerable: true,
  get: function () { return token2022.getMintCodec; }
});
Object.defineProperty(exports, "getMintDecoder", {
  enumerable: true,
  get: function () { return token2022.getMintDecoder; }
});
Object.defineProperty(exports, "getMintEncoder", {
  enumerable: true,
  get: function () { return token2022.getMintEncoder; }
});
Object.defineProperty(exports, "getMintSize", {
  enumerable: true,
  get: function () { return token2022.getMintSize; }
});
Object.defineProperty(exports, "getMintToCheckedDiscriminatorBytes", {
  enumerable: true,
  get: function () { return token2022.getMintToCheckedDiscriminatorBytes; }
});
Object.defineProperty(exports, "getMintToCheckedInstruction", {
  enumerable: true,
  get: function () { return token2022.getMintToCheckedInstruction; }
});
Object.defineProperty(exports, "getMintToCheckedInstructionDataCodec", {
  enumerable: true,
  get: function () { return token2022.getMintToCheckedInstructionDataCodec; }
});
Object.defineProperty(exports, "getMintToCheckedInstructionDataDecoder", {
  enumerable: true,
  get: function () { return token2022.getMintToCheckedInstructionDataDecoder; }
});
Object.defineProperty(exports, "getMintToCheckedInstructionDataEncoder", {
  enumerable: true,
  get: function () { return token2022.getMintToCheckedInstructionDataEncoder; }
});
Object.defineProperty(exports, "getMintToDiscriminatorBytes", {
  enumerable: true,
  get: function () { return token2022.getMintToDiscriminatorBytes; }
});
Object.defineProperty(exports, "getMintToInstruction", {
  enumerable: true,
  get: function () { return token2022.getMintToInstruction; }
});
Object.defineProperty(exports, "getMintToInstructionDataCodec", {
  enumerable: true,
  get: function () { return token2022.getMintToInstructionDataCodec; }
});
Object.defineProperty(exports, "getMintToInstructionDataDecoder", {
  enumerable: true,
  get: function () { return token2022.getMintToInstructionDataDecoder; }
});
Object.defineProperty(exports, "getMintToInstructionDataEncoder", {
  enumerable: true,
  get: function () { return token2022.getMintToInstructionDataEncoder; }
});
Object.defineProperty(exports, "getMultisigCodec", {
  enumerable: true,
  get: function () { return token2022.getMultisigCodec; }
});
Object.defineProperty(exports, "getMultisigDecoder", {
  enumerable: true,
  get: function () { return token2022.getMultisigDecoder; }
});
Object.defineProperty(exports, "getMultisigEncoder", {
  enumerable: true,
  get: function () { return token2022.getMultisigEncoder; }
});
Object.defineProperty(exports, "getMultisigSize", {
  enumerable: true,
  get: function () { return token2022.getMultisigSize; }
});
Object.defineProperty(exports, "getPauseDiscriminatorBytes", {
  enumerable: true,
  get: function () { return token2022.getPauseDiscriminatorBytes; }
});
Object.defineProperty(exports, "getPauseInstruction", {
  enumerable: true,
  get: function () { return token2022.getPauseInstruction; }
});
Object.defineProperty(exports, "getPauseInstructionDataCodec", {
  enumerable: true,
  get: function () { return token2022.getPauseInstructionDataCodec; }
});
Object.defineProperty(exports, "getPauseInstructionDataDecoder", {
  enumerable: true,
  get: function () { return token2022.getPauseInstructionDataDecoder; }
});
Object.defineProperty(exports, "getPauseInstructionDataEncoder", {
  enumerable: true,
  get: function () { return token2022.getPauseInstructionDataEncoder; }
});
Object.defineProperty(exports, "getPausePausableDiscriminatorBytes", {
  enumerable: true,
  get: function () { return token2022.getPausePausableDiscriminatorBytes; }
});
Object.defineProperty(exports, "getPostInitializeInstructionsForMintExtensions", {
  enumerable: true,
  get: function () { return token2022.getPostInitializeInstructionsForMintExtensions; }
});
Object.defineProperty(exports, "getPostInitializeInstructionsForTokenExtensions", {
  enumerable: true,
  get: function () { return token2022.getPostInitializeInstructionsForTokenExtensions; }
});
Object.defineProperty(exports, "getPreInitializeInstructionsForMintExtensions", {
  enumerable: true,
  get: function () { return token2022.getPreInitializeInstructionsForMintExtensions; }
});
Object.defineProperty(exports, "getReallocateDiscriminatorBytes", {
  enumerable: true,
  get: function () { return token2022.getReallocateDiscriminatorBytes; }
});
Object.defineProperty(exports, "getReallocateInstruction", {
  enumerable: true,
  get: function () { return token2022.getReallocateInstruction; }
});
Object.defineProperty(exports, "getReallocateInstructionDataCodec", {
  enumerable: true,
  get: function () { return token2022.getReallocateInstructionDataCodec; }
});
Object.defineProperty(exports, "getReallocateInstructionDataDecoder", {
  enumerable: true,
  get: function () { return token2022.getReallocateInstructionDataDecoder; }
});
Object.defineProperty(exports, "getReallocateInstructionDataEncoder", {
  enumerable: true,
  get: function () { return token2022.getReallocateInstructionDataEncoder; }
});
Object.defineProperty(exports, "getRecoverNestedAssociatedTokenDiscriminatorBytes", {
  enumerable: true,
  get: function () { return token2022.getRecoverNestedAssociatedTokenDiscriminatorBytes; }
});
Object.defineProperty(exports, "getRecoverNestedAssociatedTokenInstruction", {
  enumerable: true,
  get: function () { return token2022.getRecoverNestedAssociatedTokenInstruction; }
});
Object.defineProperty(exports, "getRecoverNestedAssociatedTokenInstructionAsync", {
  enumerable: true,
  get: function () { return token2022.getRecoverNestedAssociatedTokenInstructionAsync; }
});
Object.defineProperty(exports, "getRecoverNestedAssociatedTokenInstructionDataCodec", {
  enumerable: true,
  get: function () { return token2022.getRecoverNestedAssociatedTokenInstructionDataCodec; }
});
Object.defineProperty(exports, "getRecoverNestedAssociatedTokenInstructionDataDecoder", {
  enumerable: true,
  get: function () { return token2022.getRecoverNestedAssociatedTokenInstructionDataDecoder; }
});
Object.defineProperty(exports, "getRecoverNestedAssociatedTokenInstructionDataEncoder", {
  enumerable: true,
  get: function () { return token2022.getRecoverNestedAssociatedTokenInstructionDataEncoder; }
});
Object.defineProperty(exports, "getRemoveTokenMetadataKeyDiscriminatorBytes", {
  enumerable: true,
  get: function () { return token2022.getRemoveTokenMetadataKeyDiscriminatorBytes; }
});
Object.defineProperty(exports, "getRemoveTokenMetadataKeyInstruction", {
  enumerable: true,
  get: function () { return token2022.getRemoveTokenMetadataKeyInstruction; }
});
Object.defineProperty(exports, "getRemoveTokenMetadataKeyInstructionDataCodec", {
  enumerable: true,
  get: function () { return token2022.getRemoveTokenMetadataKeyInstructionDataCodec; }
});
Object.defineProperty(exports, "getRemoveTokenMetadataKeyInstructionDataDecoder", {
  enumerable: true,
  get: function () { return token2022.getRemoveTokenMetadataKeyInstructionDataDecoder; }
});
Object.defineProperty(exports, "getRemoveTokenMetadataKeyInstructionDataEncoder", {
  enumerable: true,
  get: function () { return token2022.getRemoveTokenMetadataKeyInstructionDataEncoder; }
});
Object.defineProperty(exports, "getResumeDiscriminatorBytes", {
  enumerable: true,
  get: function () { return token2022.getResumeDiscriminatorBytes; }
});
Object.defineProperty(exports, "getResumeInstruction", {
  enumerable: true,
  get: function () { return token2022.getResumeInstruction; }
});
Object.defineProperty(exports, "getResumeInstructionDataCodec", {
  enumerable: true,
  get: function () { return token2022.getResumeInstructionDataCodec; }
});
Object.defineProperty(exports, "getResumeInstructionDataDecoder", {
  enumerable: true,
  get: function () { return token2022.getResumeInstructionDataDecoder; }
});
Object.defineProperty(exports, "getResumeInstructionDataEncoder", {
  enumerable: true,
  get: function () { return token2022.getResumeInstructionDataEncoder; }
});
Object.defineProperty(exports, "getResumePausableDiscriminatorBytes", {
  enumerable: true,
  get: function () { return token2022.getResumePausableDiscriminatorBytes; }
});
Object.defineProperty(exports, "getRevokeDiscriminatorBytes", {
  enumerable: true,
  get: function () { return token2022.getRevokeDiscriminatorBytes; }
});
Object.defineProperty(exports, "getRevokeInstruction", {
  enumerable: true,
  get: function () { return token2022.getRevokeInstruction; }
});
Object.defineProperty(exports, "getRevokeInstructionDataCodec", {
  enumerable: true,
  get: function () { return token2022.getRevokeInstructionDataCodec; }
});
Object.defineProperty(exports, "getRevokeInstructionDataDecoder", {
  enumerable: true,
  get: function () { return token2022.getRevokeInstructionDataDecoder; }
});
Object.defineProperty(exports, "getRevokeInstructionDataEncoder", {
  enumerable: true,
  get: function () { return token2022.getRevokeInstructionDataEncoder; }
});
Object.defineProperty(exports, "getSetAuthorityDiscriminatorBytes", {
  enumerable: true,
  get: function () { return token2022.getSetAuthorityDiscriminatorBytes; }
});
Object.defineProperty(exports, "getSetAuthorityInstruction", {
  enumerable: true,
  get: function () { return token2022.getSetAuthorityInstruction; }
});
Object.defineProperty(exports, "getSetAuthorityInstructionDataCodec", {
  enumerable: true,
  get: function () { return token2022.getSetAuthorityInstructionDataCodec; }
});
Object.defineProperty(exports, "getSetAuthorityInstructionDataDecoder", {
  enumerable: true,
  get: function () { return token2022.getSetAuthorityInstructionDataDecoder; }
});
Object.defineProperty(exports, "getSetAuthorityInstructionDataEncoder", {
  enumerable: true,
  get: function () { return token2022.getSetAuthorityInstructionDataEncoder; }
});
Object.defineProperty(exports, "getSetTransferFeeDiscriminatorBytes", {
  enumerable: true,
  get: function () { return token2022.getSetTransferFeeDiscriminatorBytes; }
});
Object.defineProperty(exports, "getSetTransferFeeInstruction", {
  enumerable: true,
  get: function () { return token2022.getSetTransferFeeInstruction; }
});
Object.defineProperty(exports, "getSetTransferFeeInstructionDataCodec", {
  enumerable: true,
  get: function () { return token2022.getSetTransferFeeInstructionDataCodec; }
});
Object.defineProperty(exports, "getSetTransferFeeInstructionDataDecoder", {
  enumerable: true,
  get: function () { return token2022.getSetTransferFeeInstructionDataDecoder; }
});
Object.defineProperty(exports, "getSetTransferFeeInstructionDataEncoder", {
  enumerable: true,
  get: function () { return token2022.getSetTransferFeeInstructionDataEncoder; }
});
Object.defineProperty(exports, "getSetTransferFeeTransferFeeDiscriminatorBytes", {
  enumerable: true,
  get: function () { return token2022.getSetTransferFeeTransferFeeDiscriminatorBytes; }
});
Object.defineProperty(exports, "getSyncNativeDiscriminatorBytes", {
  enumerable: true,
  get: function () { return token2022.getSyncNativeDiscriminatorBytes; }
});
Object.defineProperty(exports, "getSyncNativeInstruction", {
  enumerable: true,
  get: function () { return token2022.getSyncNativeInstruction; }
});
Object.defineProperty(exports, "getSyncNativeInstructionDataCodec", {
  enumerable: true,
  get: function () { return token2022.getSyncNativeInstructionDataCodec; }
});
Object.defineProperty(exports, "getSyncNativeInstructionDataDecoder", {
  enumerable: true,
  get: function () { return token2022.getSyncNativeInstructionDataDecoder; }
});
Object.defineProperty(exports, "getSyncNativeInstructionDataEncoder", {
  enumerable: true,
  get: function () { return token2022.getSyncNativeInstructionDataEncoder; }
});
Object.defineProperty(exports, "getThawAccountDiscriminatorBytes", {
  enumerable: true,
  get: function () { return token2022.getThawAccountDiscriminatorBytes; }
});
Object.defineProperty(exports, "getThawAccountInstruction", {
  enumerable: true,
  get: function () { return token2022.getThawAccountInstruction; }
});
Object.defineProperty(exports, "getThawAccountInstructionDataCodec", {
  enumerable: true,
  get: function () { return token2022.getThawAccountInstructionDataCodec; }
});
Object.defineProperty(exports, "getThawAccountInstructionDataDecoder", {
  enumerable: true,
  get: function () { return token2022.getThawAccountInstructionDataDecoder; }
});
Object.defineProperty(exports, "getThawAccountInstructionDataEncoder", {
  enumerable: true,
  get: function () { return token2022.getThawAccountInstructionDataEncoder; }
});
Object.defineProperty(exports, "getToken2022ErrorMessage", {
  enumerable: true,
  get: function () { return token2022.getToken2022ErrorMessage; }
});
Object.defineProperty(exports, "getTokenCodec", {
  enumerable: true,
  get: function () { return token2022.getTokenCodec; }
});
Object.defineProperty(exports, "getTokenDecoder", {
  enumerable: true,
  get: function () { return token2022.getTokenDecoder; }
});
Object.defineProperty(exports, "getTokenEncoder", {
  enumerable: true,
  get: function () { return token2022.getTokenEncoder; }
});
Object.defineProperty(exports, "getTokenMetadataFieldCodec", {
  enumerable: true,
  get: function () { return token2022.getTokenMetadataFieldCodec; }
});
Object.defineProperty(exports, "getTokenMetadataFieldDecoder", {
  enumerable: true,
  get: function () { return token2022.getTokenMetadataFieldDecoder; }
});
Object.defineProperty(exports, "getTokenMetadataFieldEncoder", {
  enumerable: true,
  get: function () { return token2022.getTokenMetadataFieldEncoder; }
});
Object.defineProperty(exports, "getTokenSize", {
  enumerable: true,
  get: function () { return token2022.getTokenSize; }
});
Object.defineProperty(exports, "getTransferCheckedDiscriminatorBytes", {
  enumerable: true,
  get: function () { return token2022.getTransferCheckedDiscriminatorBytes; }
});
Object.defineProperty(exports, "getTransferCheckedInstruction", {
  enumerable: true,
  get: function () { return token2022.getTransferCheckedInstruction; }
});
Object.defineProperty(exports, "getTransferCheckedInstructionDataCodec", {
  enumerable: true,
  get: function () { return token2022.getTransferCheckedInstructionDataCodec; }
});
Object.defineProperty(exports, "getTransferCheckedInstructionDataDecoder", {
  enumerable: true,
  get: function () { return token2022.getTransferCheckedInstructionDataDecoder; }
});
Object.defineProperty(exports, "getTransferCheckedInstructionDataEncoder", {
  enumerable: true,
  get: function () { return token2022.getTransferCheckedInstructionDataEncoder; }
});
Object.defineProperty(exports, "getTransferCheckedWithFeeDiscriminatorBytes", {
  enumerable: true,
  get: function () { return token2022.getTransferCheckedWithFeeDiscriminatorBytes; }
});
Object.defineProperty(exports, "getTransferCheckedWithFeeInstruction", {
  enumerable: true,
  get: function () { return token2022.getTransferCheckedWithFeeInstruction; }
});
Object.defineProperty(exports, "getTransferCheckedWithFeeInstructionDataCodec", {
  enumerable: true,
  get: function () { return token2022.getTransferCheckedWithFeeInstructionDataCodec; }
});
Object.defineProperty(exports, "getTransferCheckedWithFeeInstructionDataDecoder", {
  enumerable: true,
  get: function () { return token2022.getTransferCheckedWithFeeInstructionDataDecoder; }
});
Object.defineProperty(exports, "getTransferCheckedWithFeeInstructionDataEncoder", {
  enumerable: true,
  get: function () { return token2022.getTransferCheckedWithFeeInstructionDataEncoder; }
});
Object.defineProperty(exports, "getTransferCheckedWithFeeTransferFeeDiscriminatorBytes", {
  enumerable: true,
  get: function () { return token2022.getTransferCheckedWithFeeTransferFeeDiscriminatorBytes; }
});
Object.defineProperty(exports, "getTransferDiscriminatorBytes", {
  enumerable: true,
  get: function () { return token2022.getTransferDiscriminatorBytes; }
});
Object.defineProperty(exports, "getTransferFeeCodec", {
  enumerable: true,
  get: function () { return token2022.getTransferFeeCodec; }
});
Object.defineProperty(exports, "getTransferFeeDecoder", {
  enumerable: true,
  get: function () { return token2022.getTransferFeeDecoder; }
});
Object.defineProperty(exports, "getTransferFeeEncoder", {
  enumerable: true,
  get: function () { return token2022.getTransferFeeEncoder; }
});
Object.defineProperty(exports, "getTransferInstruction", {
  enumerable: true,
  get: function () { return token2022.getTransferInstruction; }
});
Object.defineProperty(exports, "getTransferInstructionDataCodec", {
  enumerable: true,
  get: function () { return token2022.getTransferInstructionDataCodec; }
});
Object.defineProperty(exports, "getTransferInstructionDataDecoder", {
  enumerable: true,
  get: function () { return token2022.getTransferInstructionDataDecoder; }
});
Object.defineProperty(exports, "getTransferInstructionDataEncoder", {
  enumerable: true,
  get: function () { return token2022.getTransferInstructionDataEncoder; }
});
Object.defineProperty(exports, "getUiAmountToAmountDiscriminatorBytes", {
  enumerable: true,
  get: function () { return token2022.getUiAmountToAmountDiscriminatorBytes; }
});
Object.defineProperty(exports, "getUiAmountToAmountInstruction", {
  enumerable: true,
  get: function () { return token2022.getUiAmountToAmountInstruction; }
});
Object.defineProperty(exports, "getUiAmountToAmountInstructionDataCodec", {
  enumerable: true,
  get: function () { return token2022.getUiAmountToAmountInstructionDataCodec; }
});
Object.defineProperty(exports, "getUiAmountToAmountInstructionDataDecoder", {
  enumerable: true,
  get: function () { return token2022.getUiAmountToAmountInstructionDataDecoder; }
});
Object.defineProperty(exports, "getUiAmountToAmountInstructionDataEncoder", {
  enumerable: true,
  get: function () { return token2022.getUiAmountToAmountInstructionDataEncoder; }
});
Object.defineProperty(exports, "getUpdateConfidentialTransferMintConfidentialTransferDiscriminatorBytes", {
  enumerable: true,
  get: function () { return token2022.getUpdateConfidentialTransferMintConfidentialTransferDiscriminatorBytes; }
});
Object.defineProperty(exports, "getUpdateConfidentialTransferMintDiscriminatorBytes", {
  enumerable: true,
  get: function () { return token2022.getUpdateConfidentialTransferMintDiscriminatorBytes; }
});
Object.defineProperty(exports, "getUpdateConfidentialTransferMintInstruction", {
  enumerable: true,
  get: function () { return token2022.getUpdateConfidentialTransferMintInstruction; }
});
Object.defineProperty(exports, "getUpdateConfidentialTransferMintInstructionDataCodec", {
  enumerable: true,
  get: function () { return token2022.getUpdateConfidentialTransferMintInstructionDataCodec; }
});
Object.defineProperty(exports, "getUpdateConfidentialTransferMintInstructionDataDecoder", {
  enumerable: true,
  get: function () { return token2022.getUpdateConfidentialTransferMintInstructionDataDecoder; }
});
Object.defineProperty(exports, "getUpdateConfidentialTransferMintInstructionDataEncoder", {
  enumerable: true,
  get: function () { return token2022.getUpdateConfidentialTransferMintInstructionDataEncoder; }
});
Object.defineProperty(exports, "getUpdateDefaultAccountStateDefaultAccountStateDiscriminatorBytes", {
  enumerable: true,
  get: function () { return token2022.getUpdateDefaultAccountStateDefaultAccountStateDiscriminatorBytes; }
});
Object.defineProperty(exports, "getUpdateDefaultAccountStateDiscriminatorBytes", {
  enumerable: true,
  get: function () { return token2022.getUpdateDefaultAccountStateDiscriminatorBytes; }
});
Object.defineProperty(exports, "getUpdateDefaultAccountStateInstruction", {
  enumerable: true,
  get: function () { return token2022.getUpdateDefaultAccountStateInstruction; }
});
Object.defineProperty(exports, "getUpdateDefaultAccountStateInstructionDataCodec", {
  enumerable: true,
  get: function () { return token2022.getUpdateDefaultAccountStateInstructionDataCodec; }
});
Object.defineProperty(exports, "getUpdateDefaultAccountStateInstructionDataDecoder", {
  enumerable: true,
  get: function () { return token2022.getUpdateDefaultAccountStateInstructionDataDecoder; }
});
Object.defineProperty(exports, "getUpdateDefaultAccountStateInstructionDataEncoder", {
  enumerable: true,
  get: function () { return token2022.getUpdateDefaultAccountStateInstructionDataEncoder; }
});
Object.defineProperty(exports, "getUpdateGroupMemberPointerDiscriminatorBytes", {
  enumerable: true,
  get: function () { return token2022.getUpdateGroupMemberPointerDiscriminatorBytes; }
});
Object.defineProperty(exports, "getUpdateGroupMemberPointerGroupMemberPointerDiscriminatorBytes", {
  enumerable: true,
  get: function () { return token2022.getUpdateGroupMemberPointerGroupMemberPointerDiscriminatorBytes; }
});
Object.defineProperty(exports, "getUpdateGroupMemberPointerInstruction", {
  enumerable: true,
  get: function () { return token2022.getUpdateGroupMemberPointerInstruction; }
});
Object.defineProperty(exports, "getUpdateGroupMemberPointerInstructionDataCodec", {
  enumerable: true,
  get: function () { return token2022.getUpdateGroupMemberPointerInstructionDataCodec; }
});
Object.defineProperty(exports, "getUpdateGroupMemberPointerInstructionDataDecoder", {
  enumerable: true,
  get: function () { return token2022.getUpdateGroupMemberPointerInstructionDataDecoder; }
});
Object.defineProperty(exports, "getUpdateGroupMemberPointerInstructionDataEncoder", {
  enumerable: true,
  get: function () { return token2022.getUpdateGroupMemberPointerInstructionDataEncoder; }
});
Object.defineProperty(exports, "getUpdateGroupPointerDiscriminatorBytes", {
  enumerable: true,
  get: function () { return token2022.getUpdateGroupPointerDiscriminatorBytes; }
});
Object.defineProperty(exports, "getUpdateGroupPointerGroupPointerDiscriminatorBytes", {
  enumerable: true,
  get: function () { return token2022.getUpdateGroupPointerGroupPointerDiscriminatorBytes; }
});
Object.defineProperty(exports, "getUpdateGroupPointerInstruction", {
  enumerable: true,
  get: function () { return token2022.getUpdateGroupPointerInstruction; }
});
Object.defineProperty(exports, "getUpdateGroupPointerInstructionDataCodec", {
  enumerable: true,
  get: function () { return token2022.getUpdateGroupPointerInstructionDataCodec; }
});
Object.defineProperty(exports, "getUpdateGroupPointerInstructionDataDecoder", {
  enumerable: true,
  get: function () { return token2022.getUpdateGroupPointerInstructionDataDecoder; }
});
Object.defineProperty(exports, "getUpdateGroupPointerInstructionDataEncoder", {
  enumerable: true,
  get: function () { return token2022.getUpdateGroupPointerInstructionDataEncoder; }
});
Object.defineProperty(exports, "getUpdateMetadataPointerDiscriminatorBytes", {
  enumerable: true,
  get: function () { return token2022.getUpdateMetadataPointerDiscriminatorBytes; }
});
Object.defineProperty(exports, "getUpdateMetadataPointerInstruction", {
  enumerable: true,
  get: function () { return token2022.getUpdateMetadataPointerInstruction; }
});
Object.defineProperty(exports, "getUpdateMetadataPointerInstructionDataCodec", {
  enumerable: true,
  get: function () { return token2022.getUpdateMetadataPointerInstructionDataCodec; }
});
Object.defineProperty(exports, "getUpdateMetadataPointerInstructionDataDecoder", {
  enumerable: true,
  get: function () { return token2022.getUpdateMetadataPointerInstructionDataDecoder; }
});
Object.defineProperty(exports, "getUpdateMetadataPointerInstructionDataEncoder", {
  enumerable: true,
  get: function () { return token2022.getUpdateMetadataPointerInstructionDataEncoder; }
});
Object.defineProperty(exports, "getUpdateMetadataPointerMetadataPointerDiscriminatorBytes", {
  enumerable: true,
  get: function () { return token2022.getUpdateMetadataPointerMetadataPointerDiscriminatorBytes; }
});
Object.defineProperty(exports, "getUpdateMultiplierScaledUiMintDiscriminatorBytes", {
  enumerable: true,
  get: function () { return token2022.getUpdateMultiplierScaledUiMintDiscriminatorBytes; }
});
Object.defineProperty(exports, "getUpdateMultiplierScaledUiMintInstruction", {
  enumerable: true,
  get: function () { return token2022.getUpdateMultiplierScaledUiMintInstruction; }
});
Object.defineProperty(exports, "getUpdateMultiplierScaledUiMintInstructionDataCodec", {
  enumerable: true,
  get: function () { return token2022.getUpdateMultiplierScaledUiMintInstructionDataCodec; }
});
Object.defineProperty(exports, "getUpdateMultiplierScaledUiMintInstructionDataDecoder", {
  enumerable: true,
  get: function () { return token2022.getUpdateMultiplierScaledUiMintInstructionDataDecoder; }
});
Object.defineProperty(exports, "getUpdateMultiplierScaledUiMintInstructionDataEncoder", {
  enumerable: true,
  get: function () { return token2022.getUpdateMultiplierScaledUiMintInstructionDataEncoder; }
});
Object.defineProperty(exports, "getUpdateMultiplierScaledUiMintScaledUiAmountMintDiscriminatorBytes", {
  enumerable: true,
  get: function () { return token2022.getUpdateMultiplierScaledUiMintScaledUiAmountMintDiscriminatorBytes; }
});
Object.defineProperty(exports, "getUpdateRateInterestBearingMintDiscriminatorBytes", {
  enumerable: true,
  get: function () { return token2022.getUpdateRateInterestBearingMintDiscriminatorBytes; }
});
Object.defineProperty(exports, "getUpdateRateInterestBearingMintInstruction", {
  enumerable: true,
  get: function () { return token2022.getUpdateRateInterestBearingMintInstruction; }
});
Object.defineProperty(exports, "getUpdateRateInterestBearingMintInstructionDataCodec", {
  enumerable: true,
  get: function () { return token2022.getUpdateRateInterestBearingMintInstructionDataCodec; }
});
Object.defineProperty(exports, "getUpdateRateInterestBearingMintInstructionDataDecoder", {
  enumerable: true,
  get: function () { return token2022.getUpdateRateInterestBearingMintInstructionDataDecoder; }
});
Object.defineProperty(exports, "getUpdateRateInterestBearingMintInstructionDataEncoder", {
  enumerable: true,
  get: function () { return token2022.getUpdateRateInterestBearingMintInstructionDataEncoder; }
});
Object.defineProperty(exports, "getUpdateRateInterestBearingMintInterestBearingMintDiscriminatorBytes", {
  enumerable: true,
  get: function () { return token2022.getUpdateRateInterestBearingMintInterestBearingMintDiscriminatorBytes; }
});
Object.defineProperty(exports, "getUpdateTokenGroupMaxSizeDiscriminatorBytes", {
  enumerable: true,
  get: function () { return token2022.getUpdateTokenGroupMaxSizeDiscriminatorBytes; }
});
Object.defineProperty(exports, "getUpdateTokenGroupMaxSizeInstruction", {
  enumerable: true,
  get: function () { return token2022.getUpdateTokenGroupMaxSizeInstruction; }
});
Object.defineProperty(exports, "getUpdateTokenGroupMaxSizeInstructionDataCodec", {
  enumerable: true,
  get: function () { return token2022.getUpdateTokenGroupMaxSizeInstructionDataCodec; }
});
Object.defineProperty(exports, "getUpdateTokenGroupMaxSizeInstructionDataDecoder", {
  enumerable: true,
  get: function () { return token2022.getUpdateTokenGroupMaxSizeInstructionDataDecoder; }
});
Object.defineProperty(exports, "getUpdateTokenGroupMaxSizeInstructionDataEncoder", {
  enumerable: true,
  get: function () { return token2022.getUpdateTokenGroupMaxSizeInstructionDataEncoder; }
});
Object.defineProperty(exports, "getUpdateTokenGroupUpdateAuthorityDiscriminatorBytes", {
  enumerable: true,
  get: function () { return token2022.getUpdateTokenGroupUpdateAuthorityDiscriminatorBytes; }
});
Object.defineProperty(exports, "getUpdateTokenGroupUpdateAuthorityInstruction", {
  enumerable: true,
  get: function () { return token2022.getUpdateTokenGroupUpdateAuthorityInstruction; }
});
Object.defineProperty(exports, "getUpdateTokenGroupUpdateAuthorityInstructionDataCodec", {
  enumerable: true,
  get: function () { return token2022.getUpdateTokenGroupUpdateAuthorityInstructionDataCodec; }
});
Object.defineProperty(exports, "getUpdateTokenGroupUpdateAuthorityInstructionDataDecoder", {
  enumerable: true,
  get: function () { return token2022.getUpdateTokenGroupUpdateAuthorityInstructionDataDecoder; }
});
Object.defineProperty(exports, "getUpdateTokenGroupUpdateAuthorityInstructionDataEncoder", {
  enumerable: true,
  get: function () { return token2022.getUpdateTokenGroupUpdateAuthorityInstructionDataEncoder; }
});
Object.defineProperty(exports, "getUpdateTokenMetadataFieldDiscriminatorBytes", {
  enumerable: true,
  get: function () { return token2022.getUpdateTokenMetadataFieldDiscriminatorBytes; }
});
Object.defineProperty(exports, "getUpdateTokenMetadataFieldInstruction", {
  enumerable: true,
  get: function () { return token2022.getUpdateTokenMetadataFieldInstruction; }
});
Object.defineProperty(exports, "getUpdateTokenMetadataFieldInstructionDataCodec", {
  enumerable: true,
  get: function () { return token2022.getUpdateTokenMetadataFieldInstructionDataCodec; }
});
Object.defineProperty(exports, "getUpdateTokenMetadataFieldInstructionDataDecoder", {
  enumerable: true,
  get: function () { return token2022.getUpdateTokenMetadataFieldInstructionDataDecoder; }
});
Object.defineProperty(exports, "getUpdateTokenMetadataFieldInstructionDataEncoder", {
  enumerable: true,
  get: function () { return token2022.getUpdateTokenMetadataFieldInstructionDataEncoder; }
});
Object.defineProperty(exports, "getUpdateTokenMetadataUpdateAuthorityDiscriminatorBytes", {
  enumerable: true,
  get: function () { return token2022.getUpdateTokenMetadataUpdateAuthorityDiscriminatorBytes; }
});
Object.defineProperty(exports, "getUpdateTokenMetadataUpdateAuthorityInstruction", {
  enumerable: true,
  get: function () { return token2022.getUpdateTokenMetadataUpdateAuthorityInstruction; }
});
Object.defineProperty(exports, "getUpdateTokenMetadataUpdateAuthorityInstructionDataCodec", {
  enumerable: true,
  get: function () { return token2022.getUpdateTokenMetadataUpdateAuthorityInstructionDataCodec; }
});
Object.defineProperty(exports, "getUpdateTokenMetadataUpdateAuthorityInstructionDataDecoder", {
  enumerable: true,
  get: function () { return token2022.getUpdateTokenMetadataUpdateAuthorityInstructionDataDecoder; }
});
Object.defineProperty(exports, "getUpdateTokenMetadataUpdateAuthorityInstructionDataEncoder", {
  enumerable: true,
  get: function () { return token2022.getUpdateTokenMetadataUpdateAuthorityInstructionDataEncoder; }
});
Object.defineProperty(exports, "getUpdateTransferHookDiscriminatorBytes", {
  enumerable: true,
  get: function () { return token2022.getUpdateTransferHookDiscriminatorBytes; }
});
Object.defineProperty(exports, "getUpdateTransferHookInstruction", {
  enumerable: true,
  get: function () { return token2022.getUpdateTransferHookInstruction; }
});
Object.defineProperty(exports, "getUpdateTransferHookInstructionDataCodec", {
  enumerable: true,
  get: function () { return token2022.getUpdateTransferHookInstructionDataCodec; }
});
Object.defineProperty(exports, "getUpdateTransferHookInstructionDataDecoder", {
  enumerable: true,
  get: function () { return token2022.getUpdateTransferHookInstructionDataDecoder; }
});
Object.defineProperty(exports, "getUpdateTransferHookInstructionDataEncoder", {
  enumerable: true,
  get: function () { return token2022.getUpdateTransferHookInstructionDataEncoder; }
});
Object.defineProperty(exports, "getUpdateTransferHookTransferHookDiscriminatorBytes", {
  enumerable: true,
  get: function () { return token2022.getUpdateTransferHookTransferHookDiscriminatorBytes; }
});
Object.defineProperty(exports, "getWithdrawExcessLamportsDiscriminatorBytes", {
  enumerable: true,
  get: function () { return token2022.getWithdrawExcessLamportsDiscriminatorBytes; }
});
Object.defineProperty(exports, "getWithdrawExcessLamportsInstruction", {
  enumerable: true,
  get: function () { return token2022.getWithdrawExcessLamportsInstruction; }
});
Object.defineProperty(exports, "getWithdrawExcessLamportsInstructionDataCodec", {
  enumerable: true,
  get: function () { return token2022.getWithdrawExcessLamportsInstructionDataCodec; }
});
Object.defineProperty(exports, "getWithdrawExcessLamportsInstructionDataDecoder", {
  enumerable: true,
  get: function () { return token2022.getWithdrawExcessLamportsInstructionDataDecoder; }
});
Object.defineProperty(exports, "getWithdrawExcessLamportsInstructionDataEncoder", {
  enumerable: true,
  get: function () { return token2022.getWithdrawExcessLamportsInstructionDataEncoder; }
});
Object.defineProperty(exports, "getWithdrawWithheldTokensFromAccountsDiscriminatorBytes", {
  enumerable: true,
  get: function () { return token2022.getWithdrawWithheldTokensFromAccountsDiscriminatorBytes; }
});
Object.defineProperty(exports, "getWithdrawWithheldTokensFromAccountsForConfidentialTransferFeeConfidentialTransferFeeDiscriminatorBytes", {
  enumerable: true,
  get: function () { return token2022.getWithdrawWithheldTokensFromAccountsForConfidentialTransferFeeConfidentialTransferFeeDiscriminatorBytes; }
});
Object.defineProperty(exports, "getWithdrawWithheldTokensFromAccountsForConfidentialTransferFeeDiscriminatorBytes", {
  enumerable: true,
  get: function () { return token2022.getWithdrawWithheldTokensFromAccountsForConfidentialTransferFeeDiscriminatorBytes; }
});
Object.defineProperty(exports, "getWithdrawWithheldTokensFromAccountsForConfidentialTransferFeeInstruction", {
  enumerable: true,
  get: function () { return token2022.getWithdrawWithheldTokensFromAccountsForConfidentialTransferFeeInstruction; }
});
Object.defineProperty(exports, "getWithdrawWithheldTokensFromAccountsForConfidentialTransferFeeInstructionDataCodec", {
  enumerable: true,
  get: function () { return token2022.getWithdrawWithheldTokensFromAccountsForConfidentialTransferFeeInstructionDataCodec; }
});
Object.defineProperty(exports, "getWithdrawWithheldTokensFromAccountsForConfidentialTransferFeeInstructionDataDecoder", {
  enumerable: true,
  get: function () { return token2022.getWithdrawWithheldTokensFromAccountsForConfidentialTransferFeeInstructionDataDecoder; }
});
Object.defineProperty(exports, "getWithdrawWithheldTokensFromAccountsForConfidentialTransferFeeInstructionDataEncoder", {
  enumerable: true,
  get: function () { return token2022.getWithdrawWithheldTokensFromAccountsForConfidentialTransferFeeInstructionDataEncoder; }
});
Object.defineProperty(exports, "getWithdrawWithheldTokensFromAccountsInstruction", {
  enumerable: true,
  get: function () { return token2022.getWithdrawWithheldTokensFromAccountsInstruction; }
});
Object.defineProperty(exports, "getWithdrawWithheldTokensFromAccountsInstructionDataCodec", {
  enumerable: true,
  get: function () { return token2022.getWithdrawWithheldTokensFromAccountsInstructionDataCodec; }
});
Object.defineProperty(exports, "getWithdrawWithheldTokensFromAccountsInstructionDataDecoder", {
  enumerable: true,
  get: function () { return token2022.getWithdrawWithheldTokensFromAccountsInstructionDataDecoder; }
});
Object.defineProperty(exports, "getWithdrawWithheldTokensFromAccountsInstructionDataEncoder", {
  enumerable: true,
  get: function () { return token2022.getWithdrawWithheldTokensFromAccountsInstructionDataEncoder; }
});
Object.defineProperty(exports, "getWithdrawWithheldTokensFromAccountsTransferFeeDiscriminatorBytes", {
  enumerable: true,
  get: function () { return token2022.getWithdrawWithheldTokensFromAccountsTransferFeeDiscriminatorBytes; }
});
Object.defineProperty(exports, "getWithdrawWithheldTokensFromMintDiscriminatorBytes", {
  enumerable: true,
  get: function () { return token2022.getWithdrawWithheldTokensFromMintDiscriminatorBytes; }
});
Object.defineProperty(exports, "getWithdrawWithheldTokensFromMintForConfidentialTransferFeeConfidentialTransferFeeDiscriminatorBytes", {
  enumerable: true,
  get: function () { return token2022.getWithdrawWithheldTokensFromMintForConfidentialTransferFeeConfidentialTransferFeeDiscriminatorBytes; }
});
Object.defineProperty(exports, "getWithdrawWithheldTokensFromMintForConfidentialTransferFeeDiscriminatorBytes", {
  enumerable: true,
  get: function () { return token2022.getWithdrawWithheldTokensFromMintForConfidentialTransferFeeDiscriminatorBytes; }
});
Object.defineProperty(exports, "getWithdrawWithheldTokensFromMintForConfidentialTransferFeeInstruction", {
  enumerable: true,
  get: function () { return token2022.getWithdrawWithheldTokensFromMintForConfidentialTransferFeeInstruction; }
});
Object.defineProperty(exports, "getWithdrawWithheldTokensFromMintForConfidentialTransferFeeInstructionDataCodec", {
  enumerable: true,
  get: function () { return token2022.getWithdrawWithheldTokensFromMintForConfidentialTransferFeeInstructionDataCodec; }
});
Object.defineProperty(exports, "getWithdrawWithheldTokensFromMintForConfidentialTransferFeeInstructionDataDecoder", {
  enumerable: true,
  get: function () { return token2022.getWithdrawWithheldTokensFromMintForConfidentialTransferFeeInstructionDataDecoder; }
});
Object.defineProperty(exports, "getWithdrawWithheldTokensFromMintForConfidentialTransferFeeInstructionDataEncoder", {
  enumerable: true,
  get: function () { return token2022.getWithdrawWithheldTokensFromMintForConfidentialTransferFeeInstructionDataEncoder; }
});
Object.defineProperty(exports, "getWithdrawWithheldTokensFromMintInstruction", {
  enumerable: true,
  get: function () { return token2022.getWithdrawWithheldTokensFromMintInstruction; }
});
Object.defineProperty(exports, "getWithdrawWithheldTokensFromMintInstructionDataCodec", {
  enumerable: true,
  get: function () { return token2022.getWithdrawWithheldTokensFromMintInstructionDataCodec; }
});
Object.defineProperty(exports, "getWithdrawWithheldTokensFromMintInstructionDataDecoder", {
  enumerable: true,
  get: function () { return token2022.getWithdrawWithheldTokensFromMintInstructionDataDecoder; }
});
Object.defineProperty(exports, "getWithdrawWithheldTokensFromMintInstructionDataEncoder", {
  enumerable: true,
  get: function () { return token2022.getWithdrawWithheldTokensFromMintInstructionDataEncoder; }
});
Object.defineProperty(exports, "getWithdrawWithheldTokensFromMintTransferFeeDiscriminatorBytes", {
  enumerable: true,
  get: function () { return token2022.getWithdrawWithheldTokensFromMintTransferFeeDiscriminatorBytes; }
});
Object.defineProperty(exports, "identifyAssociatedTokenInstruction", {
  enumerable: true,
  get: function () { return token2022.identifyAssociatedTokenInstruction; }
});
Object.defineProperty(exports, "identifyToken2022Account", {
  enumerable: true,
  get: function () { return token2022.identifyToken2022Account; }
});
Object.defineProperty(exports, "identifyToken2022Instruction", {
  enumerable: true,
  get: function () { return token2022.identifyToken2022Instruction; }
});
Object.defineProperty(exports, "isAssociatedTokenError", {
  enumerable: true,
  get: function () { return token2022.isAssociatedTokenError; }
});
Object.defineProperty(exports, "isExtension", {
  enumerable: true,
  get: function () { return token2022.isExtension; }
});
Object.defineProperty(exports, "isToken2022Error", {
  enumerable: true,
  get: function () { return token2022.isToken2022Error; }
});
Object.defineProperty(exports, "isTokenMetadataField", {
  enumerable: true,
  get: function () { return token2022.isTokenMetadataField; }
});
Object.defineProperty(exports, "parseAmountToUiAmountInstruction", {
  enumerable: true,
  get: function () { return token2022.parseAmountToUiAmountInstruction; }
});
Object.defineProperty(exports, "parseApplyConfidentialPendingBalanceInstruction", {
  enumerable: true,
  get: function () { return token2022.parseApplyConfidentialPendingBalanceInstruction; }
});
Object.defineProperty(exports, "parseApproveCheckedInstruction", {
  enumerable: true,
  get: function () { return token2022.parseApproveCheckedInstruction; }
});
Object.defineProperty(exports, "parseApproveConfidentialTransferAccountInstruction", {
  enumerable: true,
  get: function () { return token2022.parseApproveConfidentialTransferAccountInstruction; }
});
Object.defineProperty(exports, "parseApproveInstruction", {
  enumerable: true,
  get: function () { return token2022.parseApproveInstruction; }
});
Object.defineProperty(exports, "parseBurnCheckedInstruction", {
  enumerable: true,
  get: function () { return token2022.parseBurnCheckedInstruction; }
});
Object.defineProperty(exports, "parseBurnInstruction", {
  enumerable: true,
  get: function () { return token2022.parseBurnInstruction; }
});
Object.defineProperty(exports, "parseCloseAccountInstruction", {
  enumerable: true,
  get: function () { return token2022.parseCloseAccountInstruction; }
});
Object.defineProperty(exports, "parseConfidentialDepositInstruction", {
  enumerable: true,
  get: function () { return token2022.parseConfidentialDepositInstruction; }
});
Object.defineProperty(exports, "parseConfidentialTransferInstruction", {
  enumerable: true,
  get: function () { return token2022.parseConfidentialTransferInstruction; }
});
Object.defineProperty(exports, "parseConfidentialTransferWithFeeInstruction", {
  enumerable: true,
  get: function () { return token2022.parseConfidentialTransferWithFeeInstruction; }
});
Object.defineProperty(exports, "parseConfidentialWithdrawInstruction", {
  enumerable: true,
  get: function () { return token2022.parseConfidentialWithdrawInstruction; }
});
Object.defineProperty(exports, "parseConfigureConfidentialTransferAccountInstruction", {
  enumerable: true,
  get: function () { return token2022.parseConfigureConfidentialTransferAccountInstruction; }
});
Object.defineProperty(exports, "parseCreateAssociatedTokenIdempotentInstruction", {
  enumerable: true,
  get: function () { return token2022.parseCreateAssociatedTokenIdempotentInstruction; }
});
Object.defineProperty(exports, "parseCreateAssociatedTokenInstruction", {
  enumerable: true,
  get: function () { return token2022.parseCreateAssociatedTokenInstruction; }
});
Object.defineProperty(exports, "parseCreateNativeMintInstruction", {
  enumerable: true,
  get: function () { return token2022.parseCreateNativeMintInstruction; }
});
Object.defineProperty(exports, "parseDisableConfidentialCreditsInstruction", {
  enumerable: true,
  get: function () { return token2022.parseDisableConfidentialCreditsInstruction; }
});
Object.defineProperty(exports, "parseDisableCpiGuardInstruction", {
  enumerable: true,
  get: function () { return token2022.parseDisableCpiGuardInstruction; }
});
Object.defineProperty(exports, "parseDisableHarvestToMintInstruction", {
  enumerable: true,
  get: function () { return token2022.parseDisableHarvestToMintInstruction; }
});
Object.defineProperty(exports, "parseDisableMemoTransfersInstruction", {
  enumerable: true,
  get: function () { return token2022.parseDisableMemoTransfersInstruction; }
});
Object.defineProperty(exports, "parseDisableNonConfidentialCreditsInstruction", {
  enumerable: true,
  get: function () { return token2022.parseDisableNonConfidentialCreditsInstruction; }
});
Object.defineProperty(exports, "parseEmitTokenMetadataInstruction", {
  enumerable: true,
  get: function () { return token2022.parseEmitTokenMetadataInstruction; }
});
Object.defineProperty(exports, "parseEmptyConfidentialTransferAccountInstruction", {
  enumerable: true,
  get: function () { return token2022.parseEmptyConfidentialTransferAccountInstruction; }
});
Object.defineProperty(exports, "parseEnableConfidentialCreditsInstruction", {
  enumerable: true,
  get: function () { return token2022.parseEnableConfidentialCreditsInstruction; }
});
Object.defineProperty(exports, "parseEnableCpiGuardInstruction", {
  enumerable: true,
  get: function () { return token2022.parseEnableCpiGuardInstruction; }
});
Object.defineProperty(exports, "parseEnableHarvestToMintInstruction", {
  enumerable: true,
  get: function () { return token2022.parseEnableHarvestToMintInstruction; }
});
Object.defineProperty(exports, "parseEnableMemoTransfersInstruction", {
  enumerable: true,
  get: function () { return token2022.parseEnableMemoTransfersInstruction; }
});
Object.defineProperty(exports, "parseEnableNonConfidentialCreditsInstruction", {
  enumerable: true,
  get: function () { return token2022.parseEnableNonConfidentialCreditsInstruction; }
});
Object.defineProperty(exports, "parseFreezeAccountInstruction", {
  enumerable: true,
  get: function () { return token2022.parseFreezeAccountInstruction; }
});
Object.defineProperty(exports, "parseGetAccountDataSizeInstruction", {
  enumerable: true,
  get: function () { return token2022.parseGetAccountDataSizeInstruction; }
});
Object.defineProperty(exports, "parseHarvestWithheldTokensToMintForConfidentialTransferFeeInstruction", {
  enumerable: true,
  get: function () { return token2022.parseHarvestWithheldTokensToMintForConfidentialTransferFeeInstruction; }
});
Object.defineProperty(exports, "parseHarvestWithheldTokensToMintInstruction", {
  enumerable: true,
  get: function () { return token2022.parseHarvestWithheldTokensToMintInstruction; }
});
Object.defineProperty(exports, "parseInitializeAccount2Instruction", {
  enumerable: true,
  get: function () { return token2022.parseInitializeAccount2Instruction; }
});
Object.defineProperty(exports, "parseInitializeAccount3Instruction", {
  enumerable: true,
  get: function () { return token2022.parseInitializeAccount3Instruction; }
});
Object.defineProperty(exports, "parseInitializeAccountInstruction", {
  enumerable: true,
  get: function () { return token2022.parseInitializeAccountInstruction; }
});
Object.defineProperty(exports, "parseInitializeConfidentialTransferFeeInstruction", {
  enumerable: true,
  get: function () { return token2022.parseInitializeConfidentialTransferFeeInstruction; }
});
Object.defineProperty(exports, "parseInitializeConfidentialTransferMintInstruction", {
  enumerable: true,
  get: function () { return token2022.parseInitializeConfidentialTransferMintInstruction; }
});
Object.defineProperty(exports, "parseInitializeDefaultAccountStateInstruction", {
  enumerable: true,
  get: function () { return token2022.parseInitializeDefaultAccountStateInstruction; }
});
Object.defineProperty(exports, "parseInitializeGroupMemberPointerInstruction", {
  enumerable: true,
  get: function () { return token2022.parseInitializeGroupMemberPointerInstruction; }
});
Object.defineProperty(exports, "parseInitializeGroupPointerInstruction", {
  enumerable: true,
  get: function () { return token2022.parseInitializeGroupPointerInstruction; }
});
Object.defineProperty(exports, "parseInitializeImmutableOwnerInstruction", {
  enumerable: true,
  get: function () { return token2022.parseInitializeImmutableOwnerInstruction; }
});
Object.defineProperty(exports, "parseInitializeInterestBearingMintInstruction", {
  enumerable: true,
  get: function () { return token2022.parseInitializeInterestBearingMintInstruction; }
});
Object.defineProperty(exports, "parseInitializeMetadataPointerInstruction", {
  enumerable: true,
  get: function () { return token2022.parseInitializeMetadataPointerInstruction; }
});
Object.defineProperty(exports, "parseInitializeMint2Instruction", {
  enumerable: true,
  get: function () { return token2022.parseInitializeMint2Instruction; }
});
Object.defineProperty(exports, "parseInitializeMintCloseAuthorityInstruction", {
  enumerable: true,
  get: function () { return token2022.parseInitializeMintCloseAuthorityInstruction; }
});
Object.defineProperty(exports, "parseInitializeMintInstruction", {
  enumerable: true,
  get: function () { return token2022.parseInitializeMintInstruction; }
});
Object.defineProperty(exports, "parseInitializeMultisig2Instruction", {
  enumerable: true,
  get: function () { return token2022.parseInitializeMultisig2Instruction; }
});
Object.defineProperty(exports, "parseInitializeMultisigInstruction", {
  enumerable: true,
  get: function () { return token2022.parseInitializeMultisigInstruction; }
});
Object.defineProperty(exports, "parseInitializeNonTransferableMintInstruction", {
  enumerable: true,
  get: function () { return token2022.parseInitializeNonTransferableMintInstruction; }
});
Object.defineProperty(exports, "parseInitializePausableConfigInstruction", {
  enumerable: true,
  get: function () { return token2022.parseInitializePausableConfigInstruction; }
});
Object.defineProperty(exports, "parseInitializePermanentDelegateInstruction", {
  enumerable: true,
  get: function () { return token2022.parseInitializePermanentDelegateInstruction; }
});
Object.defineProperty(exports, "parseInitializeScaledUiAmountMintInstruction", {
  enumerable: true,
  get: function () { return token2022.parseInitializeScaledUiAmountMintInstruction; }
});
Object.defineProperty(exports, "parseInitializeTokenGroupInstruction", {
  enumerable: true,
  get: function () { return token2022.parseInitializeTokenGroupInstruction; }
});
Object.defineProperty(exports, "parseInitializeTokenGroupMemberInstruction", {
  enumerable: true,
  get: function () { return token2022.parseInitializeTokenGroupMemberInstruction; }
});
Object.defineProperty(exports, "parseInitializeTokenMetadataInstruction", {
  enumerable: true,
  get: function () { return token2022.parseInitializeTokenMetadataInstruction; }
});
Object.defineProperty(exports, "parseInitializeTransferFeeConfigInstruction", {
  enumerable: true,
  get: function () { return token2022.parseInitializeTransferFeeConfigInstruction; }
});
Object.defineProperty(exports, "parseInitializeTransferHookInstruction", {
  enumerable: true,
  get: function () { return token2022.parseInitializeTransferHookInstruction; }
});
Object.defineProperty(exports, "parseMintToCheckedInstruction", {
  enumerable: true,
  get: function () { return token2022.parseMintToCheckedInstruction; }
});
Object.defineProperty(exports, "parseMintToInstruction", {
  enumerable: true,
  get: function () { return token2022.parseMintToInstruction; }
});
Object.defineProperty(exports, "parsePauseInstruction", {
  enumerable: true,
  get: function () { return token2022.parsePauseInstruction; }
});
Object.defineProperty(exports, "parseReallocateInstruction", {
  enumerable: true,
  get: function () { return token2022.parseReallocateInstruction; }
});
Object.defineProperty(exports, "parseRecoverNestedAssociatedTokenInstruction", {
  enumerable: true,
  get: function () { return token2022.parseRecoverNestedAssociatedTokenInstruction; }
});
Object.defineProperty(exports, "parseRemoveTokenMetadataKeyInstruction", {
  enumerable: true,
  get: function () { return token2022.parseRemoveTokenMetadataKeyInstruction; }
});
Object.defineProperty(exports, "parseResumeInstruction", {
  enumerable: true,
  get: function () { return token2022.parseResumeInstruction; }
});
Object.defineProperty(exports, "parseRevokeInstruction", {
  enumerable: true,
  get: function () { return token2022.parseRevokeInstruction; }
});
Object.defineProperty(exports, "parseSetAuthorityInstruction", {
  enumerable: true,
  get: function () { return token2022.parseSetAuthorityInstruction; }
});
Object.defineProperty(exports, "parseSetTransferFeeInstruction", {
  enumerable: true,
  get: function () { return token2022.parseSetTransferFeeInstruction; }
});
Object.defineProperty(exports, "parseSyncNativeInstruction", {
  enumerable: true,
  get: function () { return token2022.parseSyncNativeInstruction; }
});
Object.defineProperty(exports, "parseThawAccountInstruction", {
  enumerable: true,
  get: function () { return token2022.parseThawAccountInstruction; }
});
Object.defineProperty(exports, "parseTransferCheckedInstruction", {
  enumerable: true,
  get: function () { return token2022.parseTransferCheckedInstruction; }
});
Object.defineProperty(exports, "parseTransferCheckedWithFeeInstruction", {
  enumerable: true,
  get: function () { return token2022.parseTransferCheckedWithFeeInstruction; }
});
Object.defineProperty(exports, "parseTransferInstruction", {
  enumerable: true,
  get: function () { return token2022.parseTransferInstruction; }
});
Object.defineProperty(exports, "parseUiAmountToAmountInstruction", {
  enumerable: true,
  get: function () { return token2022.parseUiAmountToAmountInstruction; }
});
Object.defineProperty(exports, "parseUpdateConfidentialTransferMintInstruction", {
  enumerable: true,
  get: function () { return token2022.parseUpdateConfidentialTransferMintInstruction; }
});
Object.defineProperty(exports, "parseUpdateDefaultAccountStateInstruction", {
  enumerable: true,
  get: function () { return token2022.parseUpdateDefaultAccountStateInstruction; }
});
Object.defineProperty(exports, "parseUpdateGroupMemberPointerInstruction", {
  enumerable: true,
  get: function () { return token2022.parseUpdateGroupMemberPointerInstruction; }
});
Object.defineProperty(exports, "parseUpdateGroupPointerInstruction", {
  enumerable: true,
  get: function () { return token2022.parseUpdateGroupPointerInstruction; }
});
Object.defineProperty(exports, "parseUpdateMetadataPointerInstruction", {
  enumerable: true,
  get: function () { return token2022.parseUpdateMetadataPointerInstruction; }
});
Object.defineProperty(exports, "parseUpdateMultiplierScaledUiMintInstruction", {
  enumerable: true,
  get: function () { return token2022.parseUpdateMultiplierScaledUiMintInstruction; }
});
Object.defineProperty(exports, "parseUpdateRateInterestBearingMintInstruction", {
  enumerable: true,
  get: function () { return token2022.parseUpdateRateInterestBearingMintInstruction; }
});
Object.defineProperty(exports, "parseUpdateTokenGroupMaxSizeInstruction", {
  enumerable: true,
  get: function () { return token2022.parseUpdateTokenGroupMaxSizeInstruction; }
});
Object.defineProperty(exports, "parseUpdateTokenGroupUpdateAuthorityInstruction", {
  enumerable: true,
  get: function () { return token2022.parseUpdateTokenGroupUpdateAuthorityInstruction; }
});
Object.defineProperty(exports, "parseUpdateTokenMetadataFieldInstruction", {
  enumerable: true,
  get: function () { return token2022.parseUpdateTokenMetadataFieldInstruction; }
});
Object.defineProperty(exports, "parseUpdateTokenMetadataUpdateAuthorityInstruction", {
  enumerable: true,
  get: function () { return token2022.parseUpdateTokenMetadataUpdateAuthorityInstruction; }
});
Object.defineProperty(exports, "parseUpdateTransferHookInstruction", {
  enumerable: true,
  get: function () { return token2022.parseUpdateTransferHookInstruction; }
});
Object.defineProperty(exports, "parseWithdrawExcessLamportsInstruction", {
  enumerable: true,
  get: function () { return token2022.parseWithdrawExcessLamportsInstruction; }
});
Object.defineProperty(exports, "parseWithdrawWithheldTokensFromAccountsForConfidentialTransferFeeInstruction", {
  enumerable: true,
  get: function () { return token2022.parseWithdrawWithheldTokensFromAccountsForConfidentialTransferFeeInstruction; }
});
Object.defineProperty(exports, "parseWithdrawWithheldTokensFromAccountsInstruction", {
  enumerable: true,
  get: function () { return token2022.parseWithdrawWithheldTokensFromAccountsInstruction; }
});
Object.defineProperty(exports, "parseWithdrawWithheldTokensFromMintForConfidentialTransferFeeInstruction", {
  enumerable: true,
  get: function () { return token2022.parseWithdrawWithheldTokensFromMintForConfidentialTransferFeeInstruction; }
});
Object.defineProperty(exports, "parseWithdrawWithheldTokensFromMintInstruction", {
  enumerable: true,
  get: function () { return token2022.parseWithdrawWithheldTokensFromMintInstruction; }
});
Object.defineProperty(exports, "tokenMetadataField", {
  enumerable: true,
  get: function () { return token2022.tokenMetadataField; }
});
Object.defineProperty(exports, "uiAmountToAmountForInterestBearingMintWithoutSimulation", {
  enumerable: true,
  get: function () { return token2022.uiAmountToAmountForInterestBearingMintWithoutSimulation; }
});
Object.defineProperty(exports, "uiAmountToAmountForMintWithoutSimulation", {
  enumerable: true,
  get: function () { return token2022.uiAmountToAmountForMintWithoutSimulation; }
});
Object.defineProperty(exports, "uiAmountToAmountForScaledUiAmountMintWithoutSimulation", {
  enumerable: true,
  get: function () { return token2022.uiAmountToAmountForScaledUiAmountMintWithoutSimulation; }
});
exports.TOKEN_PROGRAM_ADDRESS = TOKEN_PROGRAM_ADDRESS;
exports.assertIsSupportedTokenProgram = assertIsSupportedTokenProgram;
exports.buildCreateTokenTransaction = buildCreateTokenTransaction;
exports.buildMintTokensTransaction = buildMintTokensTransaction;
exports.buildTransferTokensTransaction = buildTransferTokensTransaction;
exports.checkedTokenProgramAddress = checkedTokenProgramAddress;
exports.getAssociatedTokenAccountAddress = getAssociatedTokenAccountAddress;
exports.getCreateTokenInstructions = getCreateTokenInstructions;
exports.getMintTokensInstructions = getMintTokensInstructions;
exports.getTransferTokensInstructions = getTransferTokensInstructions;
exports.parseTokenProgramAddressOrMoniker = parseTokenProgramAddressOrMoniker;
exports.tokenAmountToUiAmount = tokenAmountToUiAmount;
exports.tokenUiAmountToAmount = tokenUiAmountToAmount;
